{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6646aa3b-78e5-4cff-8249-efe8feb06665",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "known_gene = pd.read_csv(\"knownGene.txt\", sep=\"\\t\", header=None)\n",
    "known_canonical = pd.read_csv(\"knownCanonical2.txt\", sep=\"\\t\", header=None)\n",
    "\n",
    "canonical_transcripts = set(known_canonical[4])\n",
    "\n",
    "filtered_known_gene = known_gene[known_gene[0].isin(canonical_transcripts)]\n",
    "\n",
    "filtered_known_gene.to_csv(\"filtered_knownGene.txt\", sep=\"\\t\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "97e33623-31b0-42c2-b637-4cba027c658f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distinct transcripts in knownGene: 254491\n",
      "Distinct transcripts in knownCanonical: 64955\n",
      "\n",
      "Number of common transcripts: 64955\n",
      "Number of transcripts only in knownGene: 189536\n",
      "Number of transcripts only in knownCanonical: 0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "known_gene = pd.read_csv(\"knownGene.txt\", sep=\"\\t\", header=None)\n",
    "known_canonical = pd.read_csv(\"knownCanonical2.txt\", sep=\"\\t\", header=None)\n",
    "\n",
    "known_gene_transcripts = known_gene[0].unique()\n",
    "known_canonical_transcripts = known_canonical[4].unique()\n",
    "\n",
    "known_gene_set = set(known_gene_transcripts)\n",
    "known_canonical_set = set(known_canonical_transcripts)\n",
    "\n",
    "print(\"Distinct transcripts in knownGene:\", len(known_gene_set))\n",
    "print(\"Distinct transcripts in knownCanonical:\", len(known_canonical_set))\n",
    "\n",
    "common_transcripts = known_gene_set.intersection(known_canonical_set)\n",
    "only_in_known_gene = known_gene_set.difference(known_canonical_set)\n",
    "only_in_known_canonical = known_canonical_set.difference(known_gene_set)\n",
    "\n",
    "print(\"\\nNumber of common transcripts:\", len(common_transcripts))\n",
    "print(\"Number of transcripts only in knownGene:\", len(only_in_known_gene))\n",
    "print(\"Number of transcripts only in knownCanonical:\", len(only_in_known_canonical))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b20c6b50-311c-4c6b-b0df-fcda38c45fc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique transcripts in filtered_knownGene.txt: 64955\n"
     ]
    }
   ],
   "source": [
    "unique_transcripts = filtered_known_gene[0].unique()\n",
    "\n",
    "num_unique_transcripts = len(unique_transcripts)\n",
    "\n",
    "print(\"Number of unique transcripts in filtered_knownGene.txt:\", num_unique_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "793084bb-de7d-460b-9cc0-9cd5c6c4aca4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distinct chromosome names:\n",
      "['chr1' 'chr10' 'chr11' 'chr12' 'chr13' 'chr14' 'chr15' 'chr16' 'chr17'\n",
      " 'chr18' 'chr19' 'chr2' 'chr20' 'chr21' 'chr22' 'chr3' 'chr4' 'chr5'\n",
      " 'chr6' 'chr7' 'chr8' 'chr9' 'chrM' 'chrMT' 'chrX' 'chrY']\n"
     ]
    }
   ],
   "source": [
    "# The chromosome column is the second column (index 1)\n",
    "distinct_chromosomes = filtered_known_gene[1].unique()\n",
    "\n",
    "# Print the list of distinct chromosome names\n",
    "print(\"Distinct chromosome names:\")\n",
    "print(distinct_chromosomes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a65e7ed-d455-4722-b020-a3c7685842a6",
   "metadata": {},
   "source": [
    "ignore the above, that was just data exploration. BELOW:Filter the knownGene dataset so that it only includes transcripts present in the knownCanonical2.txt list (the canonical transcripts).\n",
    "Further filter out non-coding RNA types like pseudo, lncRNA, and lincRNA based on the gene type.\n",
    "Exclude rows with irregular chromosomes (e.g., chrM, chrMT)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "380658b5-1148-41cc-8231-49619a943f1d",
   "metadata": {},
   "source": [
    "filter out any non coding regions :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0ef1ba35-a98e-4a88-8aed-8d754daf1bc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in the filtered file: 64926\n",
      "Distinct chromosome names after filtering:\n",
      "['chr1' 'chr10' 'chr11' 'chr12' 'chr13' 'chr14' 'chr15' 'chr16' 'chr17'\n",
      " 'chr18' 'chr19' 'chr2' 'chr20' 'chr21' 'chr22' 'chr3' 'chr4' 'chr5'\n",
      " 'chr6' 'chr7' 'chr8' 'chr9' 'chrX' 'chrY']\n",
      "     0      1      2                    3   4  5      6      7         8   9   \\\n",
      "0  chr1  11868  14409  ENST00000456328.2_1   0  +  11868  11868     25600   3   \n",
      "1  chr1  12009  13670  ENST00000450305.2_2   0  +  12009  12009  16724991   6   \n",
      "2  chr1  14695  24886  ENST00000488147.2_5   0  -  14695  14695  16724991  10   \n",
      "3  chr1  29553  31097  ENST00000473358.1_5   0  +  29553  29553     25600   3   \n",
      "4  chr1  30266  31109  ENST00000469289.1_1   0  +  30266  30266     25600   2   \n",
      "\n",
      "   ...           17    18                                  19         20  \\\n",
      "0  ...      DDX11L2  none                              lncRNA  nonCoding   \n",
      "1  ...      DDX11L1  none  transcribed_unprocessed_pseudogene     pseudo   \n",
      "2  ...       WASH7P  none              unprocessed_pseudogene     pseudo   \n",
      "3  ...  MIR1302-2HG  none                              lncRNA  nonCoding   \n",
      "4  ...  MIR1302-2HG  none                              lncRNA  nonCoding   \n",
      "\n",
      "                    21                                  22  \\\n",
      "0  havana_homo_sapiens                              lncRNA   \n",
      "1  havana_homo_sapiens  transcribed_unprocessed_pseudogene   \n",
      "2  havana_homo_sapiens              unprocessed_pseudogene   \n",
      "3  havana_homo_sapiens                              lncRNA   \n",
      "4  havana_homo_sapiens                              lncRNA   \n",
      "\n",
      "                                                  23 24                   25  \\\n",
      "0        Ensembl_canonical,basic,overlaps_pseudogene  2  canonical,basic,all   \n",
      "1                            Ensembl_canonical,basic  2  canonical,basic,all   \n",
      "2                            Ensembl_canonical,basic  2  canonical,basic,all   \n",
      "3  Ensembl_canonical,basic,dotter_confirmed,ncRNA...  2  canonical,basic,all   \n",
      "4       basic,ncRNA_host,not_best_in_genome_evidence  2            basic,all   \n",
      "\n",
      "  26  \n",
      "0  0  \n",
      "1  0  \n",
      "2  0  \n",
      "3  0  \n",
      "4  0  \n",
      "\n",
      "[5 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "known_gene_bed = pd.read_csv(\"knownGene.bed\", sep=\"\\t\", header=None)\n",
    "known_canonical = pd.read_csv(\"knownCanonical2.txt\", sep=\"\\t\", header=None)\n",
    "\n",
    "# Extract the canonical transcript IDs (5th column of knownCanonical2.txt)\n",
    "canonical_transcripts = set(known_canonical[4])\n",
    "\n",
    "# Filter knownGene.bed to include only canonical transcripts (4th column of knownGene.bed)\n",
    "filtered_known_gene = known_gene_bed[known_gene_bed[3].isin(canonical_transcripts)]\n",
    "\n",
    "# List of standard chromosomes (excluding mitochondrial and irregular chromosomes)\n",
    "standard_chromosomes = {f\"chr{i}\" for i in range(1, 23)}.union({\"chrX\", \"chrY\"})\n",
    "\n",
    "# Filter out rows with irregular chromosome types (e.g., 'chrM', 'chrMT')\n",
    "filtered_no_irregular_chromosomes = filtered_known_gene[filtered_known_gene[0].isin(standard_chromosomes)]\n",
    "\n",
    "# Save the filtered DataFrame to a new file\n",
    "filtered_no_irregular_chromosomes.to_csv(\"filtered_knownGene_noIrregular.txt\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "# Display some stats\n",
    "num_filtered_rows = filtered_no_irregular_chromosomes.shape[0]\n",
    "distinct_chromosomes = filtered_no_irregular_chromosomes[0].unique()\n",
    "\n",
    "print(f\"Number of rows in the filtered file: {num_filtered_rows}\")\n",
    "print(\"Distinct chromosome names after filtering:\")\n",
    "print(distinct_chromosomes)\n",
    "print(known_gene_bed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "50a1cc24-086f-4499-9a7d-9b420d1bfda9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in the final filtered file: 19810\n",
      "Distinct gene types remaining in the final filtered file:\n",
      "['protein_coding']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the filtered file without irregular chromosomes\n",
    "filtered_no_irregular = pd.read_csv(\"filtered_knownGene_noIrregular.txt\", sep=\"\\t\", header=None)\n",
    "\n",
    "# The gene type column appears to be the 23rd column (index 22)\n",
    "# Filter out rows where the gene type is 'pseudo' or 'lncRNA'\n",
    "# Define the list of gene types to keep\n",
    "keep_gene_types = [\n",
    "    'protein_coding']\n",
    "\n",
    "# Filter the DataFrame to keep only rows with the specified gene types\n",
    "filtered_final = filtered_no_irregular[filtered_no_irregular[22].isin(keep_gene_types)]\n",
    "\n",
    "# Save the final filtered DataFrame to a new file\n",
    "filtered_final.to_csv(\"filtered_knownGene_final.txt\", sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "# Display some statistics\n",
    "num_final_rows = filtered_final.shape[0]\n",
    "distinct_final_gene_types = filtered_final[22].unique()\n",
    "\n",
    "print(f\"Number of rows in the final filtered file: {num_final_rows}\")\n",
    "print(\"Distinct gene types remaining in the final filtered file:\")\n",
    "print(distinct_final_gene_types)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95dbc95f-8149-4578-8c97-e24e75e1be04",
   "metadata": {},
   "source": [
    "ok so i did this: \n",
    "filter the knownGene dataset so that it only includes transcripts present in the knownCanonical\n",
    "- filter out non-coding RNA types like pseudo, lncRNA, and lincRNA etc\n",
    "-Exclude rows with irregular chromosomes (chrM, chrMT).\n",
    "Number of rows in the final filtered file: 26623\n",
    "Distinct gene types remaining in the final filtered file:\n",
    "['protein_coding']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "559c8f20-242b-4712-b36b-5844a3df99a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      0       1       2                    3   4  5       6       7       8   \\\n",
      "8   chr1   65418   71585  ENST00000641515.2_5   0  +   65564   70008  789624   \n",
      "12  chr1  134900  139379    ENST00000423372.3   0  -  138529  139309  789624   \n",
      "24  chr1  367639  368634    ENST00000426406.1   0  +  367658  368597  789624   \n",
      "40  chr1  621095  622034  ENST00000332831.5_7   0  -  621095  622034  789624   \n",
      "63  chr1  859302  879954  ENST00000616016.5_6   0  +  859811  879533  789624   \n",
      "\n",
      "    9   ...          17          18              19      20  \\\n",
      "8    3  ...       OR4F5  A0A2U3U0J3  protein_coding  coding   \n",
      "12   2  ...  AL627309.1      R4GN28  protein_coding  coding   \n",
      "24   1  ...      OR4F29      Q6IEY1  protein_coding  coding   \n",
      "40   1  ...      OR4F16      Q6IEY1  protein_coding  coding   \n",
      "63  14  ...      SAMD11  A0A087WU74  protein_coding  coding   \n",
      "\n",
      "                                        21              22  \\\n",
      "8                      havana_homo_sapiens  protein_coding   \n",
      "12                                 ensembl  protein_coding   \n",
      "24               ensembl_havana_transcript  protein_coding   \n",
      "40  ensembl_havana_transcript_homo_sapiens  protein_coding   \n",
      "63                     havana_homo_sapiens  protein_coding   \n",
      "\n",
      "                                                   23 24                   25  \\\n",
      "8   CCDS,Ensembl_canonical,MANE_Select,RNA_Seq_sup...  2  canonical,basic,all   \n",
      "12                             appris_principal,basic  3  canonical,basic,all   \n",
      "24                        CCDS,appris_principal,basic  2  canonical,basic,all   \n",
      "40  CCDS,Ensembl_canonical,MANE_Select,appris_prin...  2  canonical,basic,all   \n",
      "63  CAGE_supported_TSS,CCDS,Ensembl_canonical,MANE...  2  canonical,basic,all   \n",
      "\n",
      "   26  \n",
      "8   0  \n",
      "12  0  \n",
      "24  0  \n",
      "40  0  \n",
      "63  0  \n",
      "\n",
      "[5 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "print(filtered_final.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8fcf7fb4-d395-4eaa-97d5-063aa25ce42f",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome = pysam.FastaFile(\"hg19.fa\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0f4cf2bc-e152-441e-b40e-0bb4bcc91bac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence extraction complete. Results saved to 'extracted_transcript_sequences_pybedtools.txt'.\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import pandas as pd\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import pysam\n",
    "import random\n",
    "from collections import Counter\n",
    "# Load the BED file using only the first 12 columns\n",
    "bed_file = pybedtools.BedTool(pybedtools.BedTool(\"filtered_knownGene_final.txt\").cut([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]))\n",
    "\n",
    "# Reference genome file (e.g., hg38.fa)\n",
    "reference_genome = pysam.FastaFile(\"hg19.fa\")\n",
    "\n",
    "# extract sequences using PyBedTools' built-in sequence function\n",
    "bed_with_sequences = bed_file.sequence(fi=\"hg19.fa\", s=True)\n",
    "\n",
    "#read the resulting sequences from the output file generated by PyBedTools\n",
    "sequences = []\n",
    "with open(bed_with_sequences.seqfn) as seq_file:\n",
    "    transcript_id = \"\"\n",
    "    for line in seq_file:\n",
    "        if line.startswith(\">\"):\n",
    "            transcript_id = line.strip().split()[0][1:]  \n",
    "        else:\n",
    "            sequence = line.strip()\n",
    "            sequences.append({\"transcript_id\": transcript_id, \"sequence\": sequence})\n",
    "\n",
    "# Convert to DataFrame and save\n",
    "df_sequences = pd.DataFrame(sequences)\n",
    "df_sequences.to_csv(\"extracted_transcript_sequences_pybedtools.txt\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(\"Sequence extraction complete. Results saved to 'extracted_transcript_sequences_pybedtools.txt'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0eb7b7a-3fbd-4737-a69e-61c925ccef59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t65418\t71585\tENST00000641515.2_5\t0\t+\t65564\t70008\t789624\t3\t15,54,2549,\t0,101,3618,\n",
      " chr1\t134900\t139379\tENST00000423372.3\t0\t-\t138529\t139309\t789624\t2\t902,1759,\t0,2720,\n",
      " chr1\t367639\t368634\tENST00000426406.1\t0\t+\t367658\t368597\t789624\t1\t995,\t0,\n",
      " chr1\t621095\t622034\tENST00000332831.5_7\t0\t-\t621095\t622034\t789624\t1\t939,\t0,\n",
      " chr1\t859302\t879954\tENST00000616016.5_6\t0\t+\t859811\t879533\t789624\t14\t1026,92,182,51,125,90,138,163,116,79,500,125,111,667,\t0,1999,6232,7116,11849,15117,15352,17221,18213,18487,18636,19330,19775,19985,\n",
      " chr1\t879582\t894636\tENST00000327044.7_6\t0\t-\t880073\t894620\t789624\t19\t598,90,136,114,144,102,114,112,140,189,114,111,79,91,121,132,175,153,42,\t0,854,1315,1970,2199,3928,4287,6924,7797,8209,8972,9579,9801,11720,11892,12691,12896,14726,15012,\n",
      " chr1\t895963\t901099\tENST00000338591.8_3\t0\t+\t896073\t900571\t789624\t12\t217,260,122,222,117,214,145,168,89,74,182,757,\t0,709,1045,1242,1771,2120,2525,2753,3336,3523,3765,4379,\n",
      " chr1\t901861\t911245\tENST00000379410.8_9\t0\t+\t901911\t909955\t789624\t16\t133,100,147,81,73,128,96,81,76,137,150,141,141,219,49,1424,\t0,222,3795,4039,4204,4397,4631,4842,5593,5806,6379,6704,7018,7351,7834,7960,\n",
      " chr1\t910577\t917473\tENST00000433179.4_10\t0\t-\t911551\t916409\t789624\t4\t1072,126,2293,29,\t0,1301,3683,6867,\n",
      " chr1\t934343\t935477\tENST00000304952.11_11\t0\t-\t934438\t935353\t789624\t4\t469,88,96,232,\t0,562,728,902,\n",
      " None\n"
     ]
    }
   ],
   "source": [
    "print(bed_file.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9fd370b9-ce93-46dd-aa7a-52ede1f65113",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exon and intron sequence extraction complete.\n"
     ]
    }
   ],
   "source": [
    "# Helper function to get the reverse complement of a sequence\n",
    "def reverse_complement(sequence):\n",
    "    complement = str.maketrans(\"ATCGatcg\", \"TAGCtagc\")\n",
    "    return sequence.translate(complement)[::-1]\n",
    "\n",
    "exon_sequences = []\n",
    "intron_sequences = []\n",
    "\n",
    "# Process each BED record\n",
    "for bed_record in bed_file:\n",
    "    chrom = bed_record.chrom\n",
    "    strand = bed_record.strand\n",
    "    chrom_start = int(bed_record.start)  # Transcription start\n",
    "    chrom_end = int(bed_record.end)  # Transcription end\n",
    "    exon_lengths = list(map(int, bed_record.fields[10].strip(',').split(',')))  # Exon lengths\n",
    "    exon_offsets = list(map(int, bed_record.fields[11].strip(',').split(',')))  # Exon start offsets\n",
    "\n",
    "    # Calculate absolute exon starts and ends\n",
    "    exon_starts = [chrom_start + offset for offset in exon_offsets]\n",
    "    exon_ends = [start + length for start, length in zip(exon_starts, exon_lengths)]\n",
    "\n",
    "    previous_end = chrom_start\n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        # Get the exon sequence\n",
    "        exon_seq = reference_genome.fetch(chrom, exon_start, exon_end)\n",
    "        if strand == '-':\n",
    "            exon_seq = reverse_complement(exon_seq)\n",
    "        exon_sequences.append({\"transcript_id\": bed_record.name, \"sequence\": exon_seq})\n",
    "\n",
    "        if previous_end < exon_start:\n",
    "            intron_seq = reference_genome.fetch(chrom, previous_end, exon_start)\n",
    "            if strand == '-':\n",
    "                intron_seq = reverse_complement(intron_seq)\n",
    "            intron_sequences.append({\"transcript_id\": bed_record.name, \"sequence\": intron_seq})\n",
    "\n",
    "        previous_end = exon_end\n",
    "\n",
    "    if previous_end < chrom_end:\n",
    "        intron_seq = reference_genome.fetch(chrom, previous_end, chrom_end)\n",
    "        if strand == '-':\n",
    "            intron_seq = reverse_complement(intron_seq)\n",
    "        intron_sequences.append({\"transcript_id\": bed_record.name, \"sequence\": intron_seq})\n",
    "\n",
    "df_exons = pd.DataFrame(exon_sequences)\n",
    "df_exons.to_csv(\"extracted_exon_sequences.txt\", sep=\"\\t\", index=False)\n",
    "\n",
    "df_introns = pd.DataFrame(intron_sequences)\n",
    "df_introns.to_csv(\"extracted_intron_sequences.txt\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(\"Exon and intron sequence extraction complete.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "27c870bc-5e80-4c0a-8945-d25ca98721da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 1864 rows out of 186430. Results saved to 'extracted_intron_sequences_subset.txt'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "intron_file = \"extracted_intron_sequences.txt\"\n",
    "fraction_to_extract = 0.01  # Extract the first 1/10th\n",
    "\n",
    "total_rows = sum(1 for _ in open(intron_file)) - 1  # Exclude header row\n",
    "rows_to_extract = int(total_rows * fraction_to_extract)\n",
    "\n",
    "# Read the first 1/10th of the file\n",
    "df_introns = pd.read_csv(intron_file, sep=\"\\t\", nrows=rows_to_extract)\n",
    "\n",
    "df_introns.to_csv(\"extracted_intron_sequences_subset.txt\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(f\"Extracted {rows_to_extract} rows out of {total_rows}. Results saved to 'extracted_intron_sequences_subset.txt'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f0d92c4b-9857-45e2-8456-1e2d726da5e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The canonical dataset contains an exact match for the transcript:\n",
      "    geneName chrom strand  chromStart   chromEnd  \\\n",
      "925   P11182  chr1      -   100652475  100715390   \n",
      "\n",
      "                                              exonEnds  \\\n",
      "925  100661979,100671858,100672193,100676328,100680...   \n",
      "\n",
      "                                            exonStarts  \\\n",
      "925  100652475,100671786,100672001,100676250,100680...   \n",
      "\n",
      "                                         exonStartsSet  \\\n",
      "925  {100700992, 100672001, 100696289, 100681539, 1...   \n",
      "\n",
      "                                           exonEndsSet  \n",
      "925  {100672193, 100676328, 100706441, 100701068, 1...  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "canonical_dataset_file = \"canonical_dataset_created.txt\"\n",
    "\n",
    "query_gene_name = \"P11182\"\n",
    "query_chrom = \"chr1\"\n",
    "query_strand = \"-\"\n",
    "query_chromStart = 100652475\n",
    "query_chromEnd = 100715390\n",
    "query_exonStarts = {100652475, 100671786, 100672001, 100676250, 100680373, 100681539, 100684182, 100696289, 100700992, 100706317, 100715326}\n",
    "query_exonEnds = {100661979, 100671858, 100672193, 100676328, 100680540, 100681756, 100684304, 100696471, 100701068, 100706441, 100715391}\n",
    "\n",
    "# load canonical dataset\n",
    "canonical_df = pd.read_csv(canonical_dataset_file, sep=\"\\t\", header=None, names=[\n",
    "    \"geneName\", \"chrom\", \"strand\", \"chromStart\", \"chromEnd\", \"exonEnds\", \"exonStarts\"\n",
    "])\n",
    "\n",
    "# Ensure exonStarts and exonEnds are processed into sets for comparison\n",
    "canonical_df[\"exonStartsSet\"] = canonical_df[\"exonStarts\"].apply(lambda x: set(map(int, x.split(\",\"))))\n",
    "canonical_df[\"exonEndsSet\"] = canonical_df[\"exonEnds\"].apply(lambda x: set(map(int, x.split(\",\"))))\n",
    "\n",
    "matches = canonical_df[\n",
    "    (canonical_df[\"geneName\"] == query_gene_name) &\n",
    "    (canonical_df[\"chrom\"] == query_chrom) &\n",
    "    (canonical_df[\"strand\"] == query_strand) &\n",
    "    (canonical_df[\"chromStart\"] == query_chromStart) &\n",
    "    (canonical_df[\"chromEnd\"] == query_chromEnd) &\n",
    "    (canonical_df[\"exonStartsSet\"] == query_exonStarts) &\n",
    "    (canonical_df[\"exonEndsSet\"] == query_exonEnds)\n",
    "]\n",
    "\n",
    "# Output the result\n",
    "if not matches.empty:\n",
    "    print(\"The canonical dataset contains an exact match for the transcript:\")\n",
    "    print(matches)\n",
    "else:\n",
    "    print(\"No exact match for the transcript found in the canonical dataset.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "96187269-a4f7-4546-a16a-59856b239862",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No matching headers found.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = \"/mnt/lareaulab/sdahiyat/illumina/canonical_sequence_with_flanked_Ns.txt\"\n",
    "\n",
    "query_start = 100652475 - 5001  # Adjust for flanking N's\n",
    "query_end = 100715390 + 5000\n",
    "query_chrom = \"chr1\"\n",
    "\n",
    "# Read the file and check for matching coordinates\n",
    "matches = []\n",
    "with open(file_path, \"r\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\">\"):  # Check only header lines\n",
    "            header = line.strip().lstrip(\">\")\n",
    "            chrom, region = header.split(\":\")\n",
    "            region_start, region_end = map(int, region.split(\"-\"))\n",
    "            \n",
    "            \n",
    "            if chrom == query_chrom and region_start == query_start and region_end == query_end:\n",
    "                matches.append(header)\n",
    "\n",
    "\n",
    "if matches:\n",
    "    print(\"Matching headers found:\")\n",
    "    for match in matches:\n",
    "        print(match)\n",
    "else:\n",
    "    print(\"No matching headers found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8fe91477-0665-4ca5-98b6-722fcb9a629e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_intron_frequencies(intron_sequences):\n",
    "    counts = Counter(''.join(intron_sequences))\n",
    "    total = sum(counts.values())\n",
    "    frequencies = {\n",
    "        'A': counts.get('A', 0) / total,\n",
    "        'T': counts.get('T', 0) / total,\n",
    "        'C': counts.get('C', 0) / total,\n",
    "        'G': counts.get('G', 0) / total\n",
    "    }\n",
    "    return frequencies\n",
    "\n",
    "def generate_random_sequence(length, frequencies):\n",
    "    nucleotides = ['A', 'T', 'C', 'G']\n",
    "    probs = [frequencies[nuc] for nuc in nucleotides]\n",
    "    return ''.join(random.choices(nucleotides, probs, k=length))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "450bd3d2-fd7b-4ca0-860d-e72f99a813ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to get the reverse complement of a sequence\n",
    "def reverse_complement(sequence):\n",
    "    complement = str.maketrans(\"ATCGatcg\", \"TAGCtagc\")\n",
    "    return sequence.translate(complement)[::-1]\n",
    "\n",
    "# Extract intron sequences and calculate overall intron nucleotide frequencies\n",
    "intron_sequences = []\n",
    "\n",
    "for bed_record in bed_file:\n",
    "    chrom = bed_record.chrom\n",
    "    strand = bed_record.strand\n",
    "    chrom_start = int(bed_record.start)\n",
    "    chrom_end = int(bed_record.end)\n",
    "    exon_starts = list(map(int, bed_record.fields[10].strip(',').split(',')))\n",
    "    exon_ends = list(map(int, bed_record.fields[11].strip(',').split(',')))\n",
    "\n",
    "    previous_end = chrom_start\n",
    "    for start_offset, end_offset in zip(exon_starts, exon_ends):\n",
    "        exon_start = chrom_start + start_offset\n",
    "        exon_end = chrom_start + end_offset\n",
    "\n",
    "        # Get intron sequence before the current exon\n",
    "        if previous_end < exon_start:\n",
    "            intron_seq = reference_genome.fetch(chrom, previous_end, exon_start)\n",
    "            if strand == '-':\n",
    "                intron_seq = reverse_complement(intron_seq)\n",
    "            intron_sequences.append(intron_seq)\n",
    "\n",
    "        previous_end = exon_end\n",
    "\n",
    "    # Add the last intron sequence if it exists\n",
    "    if previous_end < chrom_end:\n",
    "        intron_seq = reference_genome.fetch(chrom, previous_end, chrom_end)\n",
    "        if strand == '-':\n",
    "            intron_seq = reverse_complement(intron_seq)\n",
    "        intron_sequences.append(intron_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7b9156c3-c23a-48c2-af44-7213d523f0ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import pandas as pd\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import pysam\n",
    "import random\n",
    "from collections import Counter\n",
    "intron_frequencies = calculate_intron_frequencies(intron_sequences)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "03ec8105-acf7-4b97-968f-376c2ced0290",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical dataset preparation complete. Saved to 'canonical_dataset_spliceai.fasta'.\n"
     ]
    }
   ],
   "source": [
    "# Function to process each transcript by generating random exons\n",
    "def process_transcript(transcript_id, full_sequence, bed_record):\n",
    "    chrom = bed_record.chrom\n",
    "    strand = bed_record.strand\n",
    "    chrom_start = int(bed_record.start)\n",
    "    exon_starts = list(map(int, bed_record.fields[10].strip(',').split(',')))\n",
    "    exon_ends = list(map(int, bed_record.fields[11].strip(',').split(',')))\n",
    "\n",
    "    processed_sequence = []\n",
    "    previous_end = 0\n",
    "\n",
    "    for start_offset, end_offset in zip(exon_starts, exon_ends):\n",
    "        exon_start = chrom_start + start_offset\n",
    "        exon_end = chrom_start + end_offset\n",
    "\n",
    "        # Add the intron sequence before the exon\n",
    "        if previous_end < start_offset:\n",
    "            intron_seq = full_sequence[previous_end:start_offset]\n",
    "            processed_sequence.append(intron_seq)\n",
    "\n",
    "        exon_length = exon_end - exon_start\n",
    "        random_exon = generate_random_sequence(exon_length, intron_frequencies)\n",
    "        processed_sequence.append(random_exon)\n",
    "\n",
    "        previous_end = end_offset\n",
    "\n",
    "    if previous_end < len(full_sequence):\n",
    "        processed_sequence.append(full_sequence[previous_end:])\n",
    "\n",
    "    final_sequence = ''.join(processed_sequence)\n",
    "    padded_sequence = f\"{padding_sequence}{final_sequence}{padding_sequence}\"\n",
    "\n",
    "    fasta_header = f\">{transcript_id}|{chrom}|{strand}|{chrom_start}|{previous_end}\"\n",
    "    return fasta_header, padded_sequence\n",
    "\n",
    "with open(\"canonical_dataset_spliceai.fasta\", \"w\") as output_file:\n",
    "    for idx, row in df_sequences.iterrows():\n",
    "        transcript_id = row[\"transcript_id\"]\n",
    "        full_sequence = row[\"sequence\"]\n",
    "        bed_record = bed_file[idx]\n",
    "        header, padded_sequence = process_transcript(transcript_id, full_sequence, bed_record)\n",
    "        output_file.write(f\"{header}\\n{padded_sequence}\\n\")\n",
    "\n",
    "print(\"Canonical dataset preparation complete. Saved to 'canonical_dataset_spliceai.fasta'.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ed0136e-5104-4449-8c76-57c9bdffcee9",
   "metadata": {},
   "source": [
    "\n",
    "example row:\n",
    ">chr1:65418-71585(+)|chr1|+|65418|3618\n",
    "chr1:65418-71585(+):\n",
    "\n",
    "The transcript is located on chromosome chr1.\n",
    "The genomic range spans from 65418 to 71585.\n",
    "The transcript is on the + strand (forward strand).\n",
    "|chr1|:\n",
    "\n",
    "Confirms the chromosome (chr1).\n",
    "+:\n",
    "\n",
    "Strand orientation is positive (+).\n",
    "65418:\n",
    "\n",
    "Start position of the transcript on the chromosome.\n",
    "3618:\n",
    "\n",
    "End position of the last exon, relative to the transcript start (chromStart). This value might indicate the last position of the exon offset or an internal boundary within the transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "2a6b8a05-c5b3-4531-a46f-458cfd866adb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pysam\n",
    "\n",
    "# # Paths\n",
    "# reference_genome_path = \"/mnt/lareaulab/sdahiyat/hg19.fa\"  # Reference genome\n",
    "# canonical_fasta_path = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_spliceai.fasta\"\n",
    "# output_file_path = \"/mnt/lareaulab/sdahiyat/illumina/canonical_sequence_with_flanked_Ns.txt\"\n",
    "\n",
    "\n",
    "# reference_genome = pysam.FastaFile(reference_genome_path)\n",
    "\n",
    "# with open(canonical_fasta_path, \"r\") as fasta_file, open(output_file_path, \"w\") as output_file:\n",
    "#     for line in fasta_file:\n",
    "#         if line.startswith(\">\"):  # Header line\n",
    "#             header = line.strip().lstrip(\">\")\n",
    "#             chrom, region_strand = header.split(\"|\")[0].split(\":\")\n",
    "#             strand = region_strand[-2]  # \"+\" or \"-\"\n",
    "#             region = region_strand[:-3]  # Extract region without strand info\n",
    "#             chrom_start, chrom_end = map(int, region.split(\"-\"))\n",
    "\n",
    "#             adjusted_start = max(0, chrom_start - 5001)  # Avoid negative values\n",
    "#             adjusted_end = chrom_end + 5000\n",
    "\n",
    "#             sequence = reference_genome.fetch(chrom, chrom_start, chrom_end)\n",
    "#             if strand == \"-\":  # Reverse complement for negative strand\n",
    "#                 sequence = sequence.translate(str.maketrans(\"ATCG\", \"TAGC\"))[::-1]\n",
    "\n",
    "#             padded_sequence = f\"{'N' * 5001}{sequence}{'N' * 5000}\"\n",
    "\n",
    "#             output_file.write(f\"{chrom}:{adjusted_start}-{adjusted_end}\\t{padded_sequence}\\n\")\n",
    "\n",
    "# print(f\"Canonical sequence file with flanked 'N's created at {output_file_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d3234146-cada-4352-8943-5d95e297e031",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Invalid rows:\n",
      "                                                                                                                                                                                                                              chrom  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none                       -1,0,0,   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none                         -1,0,   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none                            0,   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none                            0,   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  0,1,0,2,2,1,1,1,2,1,2,1,0,0,   \n",
      "...                                                                                                                                                                                                                             ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none                         -1,0,   \n",
      "                                                                                                                                                                                                 none                         -1,0,   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none             -1,0,2,0,0,1,0,0,   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none            0,1,1,2,1,0,1,2,0,   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none          0,2,0,0,2,0,1,1,1,1,   \n",
      "\n",
      "                                                                                                                                                                                                      chromStart  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none       none   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none       none   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none       none   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none       none   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none       none   \n",
      "...                                                                                                                                                                                                          ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none       none   \n",
      "                                                                                                                                                                                                 none       none   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none       none   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none       none   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none       none   \n",
      "\n",
      "                                                                                                                                                                                                         chromEnd  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none       OR4F5   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none  AL627309.1   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none      OR4F29   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none      OR4F16   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none      SAMD11   \n",
      "...                                                                                                                                                                                                           ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none       SPRY3   \n",
      "                                                                                                                                                                                                 none       SPRY3   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none       VAMP7   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none        IL9R   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none      WASH6P   \n",
      "\n",
      "                                                                                                                                                                                                             name  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none  A0A2U3U0J3   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none      R4GN28   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none      Q6IEY1   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none      Q6IEY1   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  A0A087WU74   \n",
      "...                                                                                                                                                                                                           ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none      O43610   \n",
      "                                                                                                                                                                                                 none      O43610   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none  A0A024R061   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none        none   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none        none   \n",
      "\n",
      "                                                                                                                                                                                                                score  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none  protein_coding   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none  protein_coding   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none  protein_coding   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none  protein_coding   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  protein_coding   \n",
      "...                                                                                                                                                                                                               ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none  protein_coding   \n",
      "                                                                                                                                                                                                 none  protein_coding   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none  protein_coding   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none  protein_coding   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none  protein_coding   \n",
      "\n",
      "                                                                                                                                                                                                       strand  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none  coding   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none  coding   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none  coding   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none  coding   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  coding   \n",
      "...                                                                                                                                                                                                       ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none  coding   \n",
      "                                                                                                                                                                                                 none  coding   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none  coding   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none  coding   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none  coding   \n",
      "\n",
      "                                                                                                                                                                                                                                   thickStart  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none                     havana_homo_sapiens   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none                                 ensembl   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none               ensembl_havana_transcript   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none  ensembl_havana_transcript_homo_sapiens   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none                     havana_homo_sapiens   \n",
      "...                                                                                                                                                                                                                                       ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none               ensembl_havana_transcript   \n",
      "                                                                                                                                                                                                 none  ensembl_havana_transcript_homo_sapiens   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none  ensembl_havana_transcript_homo_sapiens   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none  ensembl_havana_transcript_homo_sapiens   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none                     havana_homo_sapiens   \n",
      "\n",
      "                                                                                                                                                                                                             thickEnd  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none  protein_coding   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none  protein_coding   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none  protein_coding   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none  protein_coding   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  protein_coding   \n",
      "...                                                                                                                                                                                                               ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none  protein_coding   \n",
      "                                                                                                                                                                                                 none  protein_coding   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none  protein_coding   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none  protein_coding   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none  protein_coding   \n",
      "\n",
      "                                                                                                                                                                                                                                                 itemRgb  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none  CCDS,Ensembl_canonical,MANE_Select,RNA_Seq_sup...   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none                             appris_principal,basic   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none                        CCDS,appris_principal,basic   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none  CCDS,Ensembl_canonical,MANE_Select,appris_prin...   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  CAGE_supported_TSS,CCDS,Ensembl_canonical,MANE...   \n",
      "...                                                                                                                                                                                                                                                  ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none                    CCDS,PAR,appris_principal,basic   \n",
      "                                                                                                                                                                                                 none                    CCDS,PAR,appris_principal,basic   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none         Ensembl_canonical,appris_principal_1,basic   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none  Ensembl_canonical,appris_principal_1,basic,ove...   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none  Ensembl_canonical,appris_principal_1,basic,ove...   \n",
      "\n",
      "                                                                                                                                                                                                       blockCount  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none           2   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none           3   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none           2   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none           2   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none           2   \n",
      "...                                                                                                                                                                                                           ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none           2   \n",
      "                                                                                                                                                                                                 none           2   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none           2   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none           2   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none           2   \n",
      "\n",
      "                                                                                                                                                                                                                blockSizes  \\\n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none  canonical,basic,all   \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none  canonical,basic,all   \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none  canonical,basic,all   \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none  canonical,basic,all   \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none  canonical,basic,all   \n",
      "...                                                                                                                                                                                                                    ...   \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none  canonical,basic,all   \n",
      "                                                                                                                                                                                                 none  canonical,basic,all   \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none  canonical,basic,all   \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none  canonical,basic,all   \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none  canonical,basic,all   \n",
      "\n",
      "                                                                                                                                                                                                       blockStarts  \n",
      "chr1 65418    71585    ENST00000641515.2_5 0 + 65564    70008    789624 3  15,54,2549,                                        0,101,3618,                                        uc328hnl.1 none none            0  \n",
      "     134900   139379   ENST00000423372.3   0 - 138529   139309   789624 2  902,1759,                                          0,2720,                                            uc319sre.1 none none            0  \n",
      "     367639   368634   ENST00000426406.1   0 + 367658   368597   789624 1  995,                                               0,                                                 uc319vnl.1 none none            0  \n",
      "     621095   622034   ENST00000332831.5_7 0 - 621095   622034   789624 1  939,                                               0,                                                 uc317tnl.1 none none            0  \n",
      "     859302   879954   ENST00000616016.5_6 0 + 859811   879533   789624 14 1026,92,182,51,125,90,138,163,116,79,500,125,11... 0,1999,6232,7116,11849,15117,15352,17221,18213,... uc327qga.1 none none            0  \n",
      "...                                                                                                                                                                                                            ...  \n",
      "chrY 59100479 59115127 ENST00000302805.2   0 + 59106539 59107406 789624 2  325,8694,                                          0,5954,                                            uc317ncw.1 none none            0  \n",
      "                                                                                                                                                                                                 none            0  \n",
      "     59214013 59276439 ENST00000711260.1_2 0 + 59222135 59274621 789624 8  104,155,58,138,91,68,93,1887,                      0,8113,14278,16768,19153,38469,58357,60539,        uc331nzk.1 none none            0  \n",
      "     59330366 59343488 ENST00000711286.1_1 0 + 59330430 59343080 789624 9  92,114,112,179,146,202,106,85,1002,                0,5210,5753,5981,6724,7582,8387,9827,12120,        uc331oak.1 none none            0  \n",
      "     59353631 59358045 ENST00000711272.1_1 0 + 59353631 59358045 789624 10 188,112,147,137,136,202,159,153,69,134,            0,720,1038,1362,1738,2051,2341,3159,4071,4280,     uc331nzw.1 none none            0  \n",
      "\n",
      "[19810 rows x 12 columns]\n",
      "19810\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "input_file = \"/mnt/lareaulab/sdahiyat/filtered_knownGene_final.txt\"\n",
    "columns = [\n",
    "    \"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\",\n",
    "    \"thickStart\", \"thickEnd\", \"itemRgb\", \"blockCount\", \"blockSizes\", \"blockStarts\",\n",
    "]\n",
    "df = pd.read_csv(input_file, sep=\"\\t\", header=None, names=columns)\n",
    "\n",
    "# Find problematic rows\n",
    "invalid_rows = df[~df[\"chromStart\"].apply(str.isdigit) | ~df[\"chromEnd\"].apply(str.isdigit)]\n",
    "print(\"Invalid rows:\")\n",
    "print(invalid_rows)\n",
    "print(len(invalid_rows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "74f14523-6a2a-4037-82fb-4a6395d21569",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical sequence file with flanked 'N's created at /mnt/lareaulab/sdahiyat/illumina/canonical_sequence_with_flanked_Ns.txt\n"
     ]
    }
   ],
   "source": [
    "#how to make canonical sequences file\n",
    "\n",
    "\n",
    "import pysam\n",
    "import pandas as pd\n",
    "\n",
    "# Paths\n",
    "reference_genome_path = \"/mnt/lareaulab/sdahiyat/hg19.fa\"  # Reference genome\n",
    "canonical_dataset_path = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "output_file_path = \"/mnt/lareaulab/sdahiyat/illumina/canonical_sequence_with_flanked_Ns.txt\"\n",
    "\n",
    "reference_genome = pysam.FastaFile(reference_genome_path)\n",
    "\n",
    "canonical_df = pd.read_csv(\n",
    "    canonical_dataset_path,\n",
    "    sep=\"\\t\",\n",
    "\n",
    "    header=None,\n",
    "    names=[\"geneName\", \"chrom\", \"strand\", \"chromStart\", \"chromEnd\", \"exonEnds\", \"exonStarts\"]\n",
    ")\n",
    "\n",
    "# Create the canonical sequence file with flanked 'N's\n",
    "with open(output_file_path, \"w\") as output_file:\n",
    "    for _, row in canonical_df.iterrows():\n",
    "        chrom = row[\"chrom\"]\n",
    "        chrom_start = row[\"chromStart\"]\n",
    "        chrom_end = row[\"chromEnd\"]\n",
    "        strand = row[\"strand\"]\n",
    "\n",
    "        adjusted_start = max(0, chrom_start-5001)  #  or chrom_start-5001\n",
    "        adjusted_end = chrom_end +5000 #or add 5000\n",
    "\n",
    "        core_sequence = reference_genome.fetch(chrom, chrom_start-1, chrom_end)  # Use adjusted coordinates\n",
    "        # if strand == \"-\":  # Reverse complement for negative strand\n",
    "        #      core_sequence = core_sequence.translate(str.maketrans(\"ATCG\", \"TAGC\"))[::-1]\n",
    "\n",
    "        padded_sequence = f\"{'N' * 5000}{core_sequence}{'N' * 5000}\"\n",
    "\n",
    "        output_file.write(f\"{chrom}:{adjusted_start}-{adjusted_end}\\t{padded_sequence}\\n\")\n",
    "\n",
    "print(f\"Canonical sequence file with flanked 'N's created at {output_file_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e6051994-b69a-4c0a-9676-ed3843195670",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datasets in the file: []\n"
     ]
    }
   ],
   "source": [
    "import h5py\n",
    "\n",
    "file_path = \"/mnt/lareaulab/sdahiyat/illumina/datafile_train_0.h5\"\n",
    "\n",
    "with h5py.File(file_path, \"r\") as h5f:\n",
    "    print(\"Datasets in the file:\", list(h5f.keys()))\n",
    "    \n",
    "    # Inspect contents of each dataset\n",
    "    for dataset_name in h5f.keys():\n",
    "        data = h5f[dataset_name][:]\n",
    "        print(f\"\\nDataset: {dataset_name}\")\n",
    "        print(f\"Data: {data[:10]}\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "492e0dd6-27dc-4638-acd3-b9ff606a0f3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First few rows of the dataset:\n",
      "     0       1       2                    3   4  5       6       7       8   \\\n",
      "0  chr1   65418   71585  ENST00000641515.2_5   0  +   65564   70008  789624   \n",
      "1  chr1  134900  139379    ENST00000423372.3   0  -  138529  139309  789624   \n",
      "2  chr1  367639  368634    ENST00000426406.1   0  +  367658  368597  789624   \n",
      "3  chr1  621095  622034  ENST00000332831.5_7   0  -  621095  622034  789624   \n",
      "4  chr1  859302  879954  ENST00000616016.5_6   0  +  859811  879533  789624   \n",
      "\n",
      "   9   ...          17          18              19      20  \\\n",
      "0   3  ...       OR4F5  A0A2U3U0J3  protein_coding  coding   \n",
      "1   2  ...  AL627309.1      R4GN28  protein_coding  coding   \n",
      "2   1  ...      OR4F29      Q6IEY1  protein_coding  coding   \n",
      "3   1  ...      OR4F16      Q6IEY1  protein_coding  coding   \n",
      "4  14  ...      SAMD11  A0A087WU74  protein_coding  coding   \n",
      "\n",
      "                                       21              22  \\\n",
      "0                     havana_homo_sapiens  protein_coding   \n",
      "1                                 ensembl  protein_coding   \n",
      "2               ensembl_havana_transcript  protein_coding   \n",
      "3  ensembl_havana_transcript_homo_sapiens  protein_coding   \n",
      "4                     havana_homo_sapiens  protein_coding   \n",
      "\n",
      "                                                  23 24                   25  \\\n",
      "0  CCDS,Ensembl_canonical,MANE_Select,RNA_Seq_sup...  2  canonical,basic,all   \n",
      "1                             appris_principal,basic  3  canonical,basic,all   \n",
      "2                        CCDS,appris_principal,basic  2  canonical,basic,all   \n",
      "3  CCDS,Ensembl_canonical,MANE_Select,appris_prin...  2  canonical,basic,all   \n",
      "4  CAGE_supported_TSS,CCDS,Ensembl_canonical,MANE...  2  canonical,basic,all   \n",
      "\n",
      "  26  \n",
      "0  0  \n",
      "1  0  \n",
      "2  0  \n",
      "3  0  \n",
      "4  0  \n",
      "\n",
      "[5 rows x 27 columns]\n",
      "\n",
      "Dataset shape: (19810, 27)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = \"/mnt/lareaulab/sdahiyat/filtered_knownGene_final.txt\"\n",
    "df = pd.read_csv(file_path, sep=\"\\t\", header=None)\n",
    "print(\"First few rows of the dataset:\")\n",
    "print(df.head())\n",
    "print(\"\\nDataset shape:\", df.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ab1d6285-4277-4e7c-b56a-9a67034b77da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical dataset saved to illumina/canonical_dataset_created.txt\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "input_file = \"filtered_knownGene_final.txt\"\n",
    "output_file = \"illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "df = pd.read_csv(input_file, sep=\"\\t\", header=None)\n",
    "\n",
    "df.columns = [\n",
    "    \"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\",\n",
    "    \"thickStart\", \"thickEnd\", \"itemRgb\", \"blockCount\",\n",
    "    \"blockSizes\", \"blockStarts\", \"other1\", \"other2\", \"other3\",\n",
    "    \"other4\", \"other5\", \"other6\", \"geneName\", \"other9\",\n",
    "    \"geneType\", \"typeCategory\", \"extraInfo\", \"category\",\n",
    "    \"canonicalStatus\", \"placeholder\", \"additional\"\n",
    "]\n",
    "\n",
    "df[\"chromStart\"] = pd.to_numeric(df[\"chromStart\"], errors=\"coerce\")\n",
    "df[\"chromEnd\"] = pd.to_numeric(df[\"chromEnd\"], errors=\"coerce\")\n",
    "df[\"blockSizes\"] = df[\"blockSizes\"].str.rstrip(\",\").str.split(\",\").apply(lambda x: list(map(int, x)))\n",
    "df[\"blockStarts\"] = df[\"blockStarts\"].str.rstrip(\",\").str.split(\",\").apply(lambda x: list(map(int, x)))\n",
    "\n",
    "results = []\n",
    "\n",
    "for _, row in df.iterrows():\n",
    "    try:\n",
    "        chrom = row[\"chrom\"]\n",
    "        strand = row[\"strand\"]\n",
    "        gene_name = row[\"geneName\"]\n",
    "        chrom_start = row[\"chromStart\"] + 1\n",
    "        chrom_end = row[\"chromEnd\"]\n",
    "        block_sizes = row[\"blockSizes\"]\n",
    "        block_starts = row[\"blockStarts\"]\n",
    "\n",
    "        exon_ends = [chrom_start + start + size - 1 for start, size in zip(block_starts, block_sizes)]  # Adjusting exon ends\n",
    "        exon_starts = [chrom_start + start for start in block_starts]\n",
    "        exon_ends = exon_ends[:-1] \n",
    "        exon_starts = exon_starts[1:] \n",
    "        if len(exon_ends) != len(exon_starts):\n",
    "            print(f\"Warning: Unequal exon ends and starts for gene {gene_name}. Skipping.\")\n",
    "            continue\n",
    "\n",
    "        exon_ends_str = \",\".join(map(str, exon_ends))\n",
    "        exon_starts_str = \",\".join(map(str, exon_starts))\n",
    "\n",
    "        results.append([gene_name,0, chrom, strand, chrom_start, chrom_end, exon_ends_str, exon_starts_str])\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing row: {row}\")\n",
    "        print(e)\n",
    "\n",
    "output_df = pd.DataFrame(results, columns=[\"geneName\",\"paralog\", \"chrom\", \"strand\", \"chromStart\", \"chromEnd\", \"exonEnds\", \"exonStarts\"])\n",
    "\n",
    "output_df.to_csv(output_file, sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "print(f\"Canonical dataset saved to {output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d0608f74-e1a1-4654-9a2b-6717bceba2ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datasets in the file:\n"
     ]
    }
   ],
   "source": [
    "import h5py\n",
    "\n",
    "datafile_path = './illumina/datafile_train_all.h5'  # Adjust the path if needed\n",
    "\n",
    "with h5py.File(datafile_path, 'r') as h5f:\n",
    "    print(\"Datasets in the file:\")\n",
    "    for key in h5f.keys():\n",
    "        print(f\"{key}: {h5f[key].dtype}, shape={h5f[key].shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "0b8e48ab-71bb-4d6a-87c5-d14b0bc6a0c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query Gene: DBT\n",
      "geneName chrom strand  chromStart  chromEnd                                                                                                      exonEnds                                                                                                    exonStarts\n",
      "  P11182  chr1      -   100652475 100715390 100661979,100671858,100672193,100676328,100680540,100681756,100684304,100696471,100701068,100706441,100715391 100652475,100671786,100672001,100676250,100680373,100681539,100684182,100696289,100700992,100706317,100715326\n",
      "\n",
      "\n",
      "Query Gene: SLC45A1\n",
      "geneName chrom strand  chromStart  chromEnd                                                                exonEnds                                                              exonStarts\n",
      "  Q9Y2W3  chr1      +     8378174   8404225 8378247,8384787,8385451,8386103,8390997,8395651,8398053,8399759,8404226 8378174,8384366,8385358,8385878,8390269,8395497,8397876,8399553,8403807\n",
      "\n",
      "\n",
      "Query Gene: NECAP2\n",
      "geneName chrom strand  chromStart  chromEnd                                                                exonEnds                                                              exonStarts\n",
      "  Q9NVZ3  chr1      +    16767219  16786573 16767349,16770228,16774470,16774637,16775697,16778511,16782389,16786574 16767219,16770127,16774365,16774555,16775588,16778333,16782313,16785337\n",
      "\n",
      "\n",
      "Matches saved to matched_rows_by_txend.txt\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "created_file = \"canonical_dataset_created.txt\"\n",
    "query_data = \t[\n",
    "    {\"txEnd\": 100715390, \"geneName\": \"DBT\"},\n",
    "    {\"txEnd\": 8404225, \"geneName\": \"SLC45A1\"},\n",
    "    {\"txEnd\": 16786573, \"geneName\": \"NECAP2\"},\n",
    "]\n",
    "\n",
    "columns_created = [\"geneName\", \"chrom\", \"strand\", \"chromStart\", \"chromEnd\", \"exonEnds\", \"exonStarts\"]\n",
    "df_created = pd.read_csv(created_file, sep=\"\\t\", header=None, names=columns_created)\n",
    "\n",
    "df_created[\"chromEnd\"] = pd.to_numeric(df_created[\"chromEnd\"], errors=\"coerce\")\n",
    "\n",
    "matches = []\n",
    "for query in query_data:\n",
    "    tx_end, gene_name = query[\"txEnd\"], query[\"geneName\"]\n",
    "    match = df_created[df_created[\"chromEnd\"] == tx_end]\n",
    "    if not match.empty:\n",
    "        matches.append({\"queryGene\": gene_name, \"matchedRows\": match})\n",
    "\n",
    "for match in matches:\n",
    "    print(f\"Query Gene: {match['queryGene']}\")\n",
    "    print(match[\"matchedRows\"].to_string(index=False))\n",
    "    print(\"\\n\")\n",
    "\n",
    "output_file = \"matched_rows_by_txend.txt\"\n",
    "with open(output_file, \"w\") as f:\n",
    "    for match in matches:\n",
    "        f.write(f\"Query Gene: {match['queryGene']}\\n\")\n",
    "        f.write(match[\"matchedRows\"].to_string(index=False))\n",
    "        f.write(\"\\n\\n\")\n",
    "\n",
    "print(f\"Matches saved to {output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1b6f2f43-3421-4532-af3c-2a91f295c5b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A0A2U3U0J3' 'R4GN28' 'Q6IEY1' ... 'Q9NR90' 'A0A804HKV4' 'A0A024R061']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "filtered_known_gene = pd.read_csv(\"filtered_knownGene_final.txt\", sep=\"\\t\", header=None)\n",
    "gene_names = filtered_known_gene.iloc[:, 18].unique()  # Assuming 'geneName' is column 19\n",
    "print(gene_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f57fada1-5a05-4fc4-aa57-893b64d69847",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "      <th>25</th>\n",
       "      <th>26</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>71585</td>\n",
       "      <td>ENST00000641515.2_5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>65564</td>\n",
       "      <td>70008</td>\n",
       "      <td>789624</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>A0A2U3U0J3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,Ensembl_canonical,MANE_Select,RNA_Seq_sup...</td>\n",
       "      <td>2</td>\n",
       "      <td>canonical,basic,all</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0      1      2                    3   4  5      6      7       8   9   \\\n",
       "0  chr1  65418  71585  ENST00000641515.2_5   0  +  65564  70008  789624   3   \n",
       "\n",
       "   ...     17          18              19      20                   21  \\\n",
       "0  ...  OR4F5  A0A2U3U0J3  protein_coding  coding  havana_homo_sapiens   \n",
       "\n",
       "               22                                                 23 24  \\\n",
       "0  protein_coding  CCDS,Ensembl_canonical,MANE_Select,RNA_Seq_sup...  2   \n",
       "\n",
       "                    25 26  \n",
       "0  canonical,basic,all  0  \n",
       "\n",
       "[1 rows x 27 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_known_gene[filtered_known_gene.iloc[:, 18] == \"A0A2U3U0J3\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a3ae90b1-5607-4259-8745-aab2a2abf8d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical Dataset Entry:\n",
      "            0     1  2          3          4          5          6\n",
      "2      Q6IEY1  chr1  +     367639     368634     368634     367639\n",
      "3      Q6IEY1  chr1  -     621095     622034     622034     621095\n",
      "15442  Q6IEY1  chr5  +  180794268  180795263  180795263  180794268\n",
      "\n",
      "Filtered KnownGene Entry:\n",
      "      chrom  chromStart   chromEnd                 name  score strand  \\\n",
      "2      chr1      367639     368634    ENST00000426406.1      0      +   \n",
      "3      chr1      621095     622034  ENST00000332831.5_7      0      -   \n",
      "15442  chr5   180794268  180795263  ENST00000456475.1_3      0      +   \n",
      "\n",
      "       thickStart   thickEnd  itemRgb  blockCount  ...  other6 geneName  \\\n",
      "2          367658     368597   789624           1  ...  OR4F29   Q6IEY1   \n",
      "3          621095     622034   789624           1  ...  OR4F16   Q6IEY1   \n",
      "15442   180794287  180795226   789624           1  ...   OR4F3   Q6IEY1   \n",
      "\n",
      "               other9 geneType                            typeCategory  \\\n",
      "2      protein_coding   coding               ensembl_havana_transcript   \n",
      "3      protein_coding   coding  ensembl_havana_transcript_homo_sapiens   \n",
      "15442  protein_coding   coding  ensembl_havana_transcript_homo_sapiens   \n",
      "\n",
      "            extraInfo                                           category  \\\n",
      "2      protein_coding                        CCDS,appris_principal,basic   \n",
      "3      protein_coding  CCDS,Ensembl_canonical,MANE_Select,appris_prin...   \n",
      "15442  protein_coding  CCDS,Ensembl_canonical,MANE_Select,appris_prin...   \n",
      "\n",
      "      canonicalStatus          placeholder additional  \n",
      "2                   2  canonical,basic,all          0  \n",
      "3                   2  canonical,basic,all          0  \n",
      "15442               2  canonical,basic,all          0  \n",
      "\n",
      "[3 rows x 27 columns]\n",
      "\n",
      "Converted Coordinates (1-based):\n",
      "      chrom  chromStart_1based   chromEnd strand\n",
      "2      chr1             367640     368634      +\n",
      "3      chr1             621096     622034      -\n",
      "15442  chr5          180794269  180795263      +\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1318/125858280.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  known_gene_row[\"chromStart_1based\"] = known_gene_row[\"chromStart\"] + 1\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "canonical_dataset = pd.read_csv(\"illumina/canonical_dataset_created.txt\", sep=\"\\t\", header=None)\n",
    "filtered_known_gene = pd.read_csv(\"filtered_knownGene_final.txt\", sep=\"\\t\", header=None)\n",
    "\n",
    "filtered_known_gene.columns = [\n",
    "    \"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\",\n",
    "    \"thickStart\", \"thickEnd\", \"itemRgb\", \"blockCount\",\n",
    "    \"blockSizes\", \"blockStarts\", \"other1\", \"other2\", \"other3\",\n",
    "    \"other4\", \"other5\", \"other6\", \"geneName\", \"other9\",\n",
    "    \"geneType\", \"typeCategory\", \"extraInfo\", \"category\",\n",
    "    \"canonicalStatus\", \"placeholder\", \"additional\"\n",
    "]\n",
    "\n",
    "gene_to_verify = \"Q6IEY1\"  # Replace with desired gene name\n",
    "canonical_row = canonical_dataset[canonical_dataset[0] == gene_to_verify]\n",
    "if canonical_row.empty:\n",
    "    print(f\"Gene {gene_to_verify} not found in canonical_dataset_created.txt\")\n",
    "else:\n",
    "    print(\"Canonical Dataset Entry:\")\n",
    "    print(canonical_row)\n",
    "\n",
    "    # Find the same gene in filtered_knownGene_final.txt\n",
    "    known_gene_row = filtered_known_gene[filtered_known_gene[\"geneName\"] == gene_to_verify]\n",
    "    if known_gene_row.empty:\n",
    "        print(f\"Gene {gene_to_verify} not found in filtered_knownGene_final.txt\")\n",
    "    else:\n",
    "        print(\"\\nFiltered KnownGene Entry:\")\n",
    "        print(known_gene_row)\n",
    "\n",
    "        #convert coordinates for comparison\n",
    "        known_gene_row[\"chromStart_1based\"] = known_gene_row[\"chromStart\"] + 1\n",
    "        print(\"\\nConverted Coordinates (1-based):\")\n",
    "        print(known_gene_row[[\"chrom\", \"chromStart_1based\", \"chromEnd\", \"strand\"]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e008a5de-8f4a-4569-add2-2c1e4457a8b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reformatted file created at illumina/shuffled_sequences_with_flanked_Ns_reformatted.txt.\n"
     ]
    }
   ],
   "source": [
    "input_file = \"illumina/shuffled_sequences_with_flanked_Ns.txt\"\n",
    "output_file = \"illumina/shuffled_sequences_with_flanked_Ns_reformatted.txt\"\n",
    "\n",
    "with open(input_file, \"r\") as infile, open(output_file, \"w\") as outfile:\n",
    "    chrom_info = None\n",
    "    sequence = []\n",
    "\n",
    "    for line in infile:\n",
    "        line = line.strip()\n",
    "        if line.startswith(\">\"):  \n",
    "            if chrom_info and sequence:\n",
    "                outfile.write(f\"{chrom_info}\\t{''.join(sequence)}\\n\")\n",
    "            \n",
    "            chrom_info = line[1:]  \n",
    "            sequence = []  \n",
    "        else: \n",
    "            sequence.append(line.strip())\n",
    "    \n",
    "    if chrom_info and sequence:\n",
    "        outfile.write(f\"{chrom_info}\\t{''.join(sequence)}\\n\")\n",
    "\n",
    "print(f\"Reformatted file created at {output_file}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b55476f-5143-4b57-92f2-beb08dac32b2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (september)",
   "language": "python",
   "name": "september"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
