{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "84d82c69-bee9-497b-b51a-80b8dfe4913b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## A very important aspect of this project is making sure all the datasets are aligned with the hypotheses we are testing. I coded this test to ensure that the introns remained unshuffled and the exons were shuffled as expected. I always crosscheck the outputted intronic and exonic sequences with UCSC Genome Browswer. Paste the 'chr x: txstart-txend' header into the browser, and it will output the canonical sequence : https://genome.ucsc.edu/cgi-bin/hgc?hgsid=2527996562_ZzJQQA4gTz8Jemae9meIu3kk5aZ8&o=895963&g=getDna&i=mixed&c=chr1&l=895963&r=901099&db=hg19"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5df93d0-0e7d-40a5-a3bc-c73ae089ed69",
   "metadata": {},
   "source": [
    "## Immediately below is the sanity check for unshuffled sequences from which all other shuffled sequences are created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dfc609f0-95de-4883-a89d-64fa865a0e1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debugging Adjusted Junctions for chr1:901862-911245\n",
      "Adjusted Exon Starts: [132, 321, 3941, 4119, 4276, 4524, 4726, 4922, 5668, 5942, 6528, 6844, 7158, 7569, 7882]\n",
      "Adjusted Exon Ends: [222, 3795, 4039, 4204, 4397, 4631, 4842, 5593, 5806, 6379, 6704, 7018, 7351, 7834, 7960]\n",
      "Sequence Length: 9384\n",
      "Exon and Intron Sequences       Type                                           Sequence\n",
      "0     Exon  GGTGAGCGGGGCGTGGGTGCGGCCACCTGGGCGCAGGGCTCCCCCA...\n",
      "1     Exon  GGTGAGCGCGGCGTGCACGGTGGCTGTGGTCTGGGAGCGTGGCTCT...\n",
      "2     Exon  GGTGGGGGCCGGGCTGGGTGGAGCACGCTAAGGGTGCAGCATCCCC...\n",
      "3     Exon  GGTGAGGCGGTGGGCAATGGGGTGGGGCCATGGCCGCCCTTCCCTC...\n",
      "4     Exon  GGTGTTTGGGATGCTTCCCGGGCCCCCAGAGGCACTCCTGACCCAG...\n",
      "5     Exon  GGTCAGTGCCGGGGACCCCACCCCCCTCCCCACCCTGATCCTCGCA...\n",
      "6     Exon  GGTGGGTGGGAGGTGCGTGGGGCTGTAGGGGGATGGGAGGGGTGCA...\n",
      "7     Exon  GGTGGGCCCCTCCCCACTGTGGGCCCGCCCCAGGGAGGCAGCTGTG...\n",
      "8     Exon  GGTAGGGCCCTGACCCTGGTTCTGCCTCCCGCCTGGCCAGGCCATG...\n",
      "9     Exon  GGTGAGGGGTCAATAGGCCCCACAGCCCAGGTCCTGGGCAGTGGTA...\n",
      "10    Exon  GGTGAGCGCTTACGGGGTGGCAGACGAAAGTGGGGCAGAAGGCTGT...\n",
      "11    Exon  GGTGGGCCCAGCACACCCACACAGCCCCTGGCCTGGTTCCCACCGT...\n",
      "12    Exon  GGTCAGGCCCCTGCTGGGTGACAGAAAGGGTGGGAGGTGCCTGCAG...\n",
      "13    Exon  TGTGAGTAGCAGCCCCCACGCCCGTGTGCCCCGGGCTCCGGGCTGG...\n",
      "14    Exon  GGTGAGTAAGGATCCTGCCTCCTGAGGTGAGTGCCTGTTGCCTCCC...\n",
      "15  Intron  AGGAGGCTGTGGACAGGGACCCAGACTTGCCGACCTGTACGACTCT...\n",
      "16  Intron  AGAGGACAGCGCGCGGATGTCGGCCGGCCTGCCGGGCCCCGAGGCT...\n",
      "17  Intron  GACATCCTGGACCTGGAGAACCAGCGAGAAAACCTGGAGCAGCCAT...\n",
      "18  Intron  GATGTCAGCGACTGCTACCTGGAGCTATTCCCCGCCCACCTGTACT...\n",
      "19  Intron  GGGCTGTTACCGCTGACGGAGCTGAGTGTCTGCCCGCTCGAGGGGT...\n",
      "20  Intron  GCCCACTGCCCGCACCCCTCCTGGTGCTCTGCCCCAGCCGGGCCGA...\n",
      "21  Intron  CGCCGTCTAACCCGGCTGCGGACGGCGTCAGGGCACGAACCCGGCG...\n",
      "22  Intron  GAGCAGTGGGACCGGCTCTTGGTCCTGTACCCAACGTCCTTGGCCA...\n",
      "23  Intron  GGGGAGCTCCCACTCCGTGCCGTCCACATCAACCTGGAGGAGAAGG...\n",
      "24  Intron  GCCCCCTCATCAACACCATCCGCGTGGTGTGCGCCAGCTACGAGGA...\n",
      "25  Intron  GTTATGGGCAGTGGCCGAGGCTCACTCTCCTCAGGCGGACAGACCA...\n",
      "26  Intron  GACCAGGCCAACTCTGACCGTGCCAGCATTGGCCGACGGAGGACCG...\n",
      "27  Intron  CTGCACAGGCTGAGCCTGGAGAGCAGCCCAGATGCCCCTGACCACA...\n",
      "28  Intron  TTCCTCAGTGCCATGCAGAGTGCACGTGGACCCACGCCCTCGAGCC...\n",
      "29  Intron   GTCTCCTCTGCCAGGGAAGGTTCGCCCGAACCCTGGCTGCCTCTGACA\n",
      "30  Intron  ATGGTCGGTCCCCCAGGAGGAGCCGGGACCCCGGCTACGACCACCT...\n"
     ]
    }
   ],
   "source": [
    "shuffled_codons_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_sequence_unflanked_unshuffled.txt\"\n",
    "canonical_dataset_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "\n",
    "sequences = {}\n",
    "with open(shuffled_codons_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue  \n",
    "        try:\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "        except ValueError:\n",
    "            print(f\"Skipping malformed line: {line}\")\n",
    "\n",
    "junctions = {}\n",
    "with open(canonical_dataset_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if len(fields) < 8: \n",
    "            continue\n",
    "\n",
    "        chrom = fields[2]  # Chromosome\n",
    "        start = int(fields[4])  # Start position\n",
    "        end = int(fields[5])    # End position\n",
    "        exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "        exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "\n",
    "        transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "        junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "\n",
    "if target_transcript in sequences and target_transcript in junctions:\n",
    "    full_sequence = sequences[target_transcript]  # gets shuffled sequence\n",
    "    exon_starts, exon_ends = junctions[target_transcript]  #gets junctions\n",
    "    seq_start = int(target_transcript.split(\":\")[1].split(\"-\")[0])  \n",
    "    exon_starts = [s - seq_start for s in exon_starts]\n",
    "    exon_ends = [e - seq_start for e in exon_ends]\n",
    "\n",
    "    print(f\"Debugging Adjusted Junctions for {target_transcript}\")\n",
    "    print(f\"Adjusted Exon Starts: {exon_starts}\")\n",
    "    print(f\"Adjusted Exon Ends: {exon_ends}\")\n",
    "    print(f\"Sequence Length: {len(full_sequence)}\")\n",
    "\n",
    "    exons = []\n",
    "    introns = []\n",
    "    last_pos = 0  \n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        \n",
    "        if 0 <= exon_start < len(full_sequence) and 0 <= exon_end <= len(full_sequence):\n",
    "            introns.append(full_sequence[last_pos:exon_start]) \n",
    "            exons.append(full_sequence[exon_start:exon_end]) \n",
    "        else:\n",
    "            print(f\" Warning: Adjusted Exon range ({exon_start}-{exon_end}) out of bounds.\")\n",
    "        \n",
    "        last_pos = exon_end\n",
    "\n",
    "    introns.append(full_sequence[last_pos:])  \n",
    "\n",
    "    import pandas as pd\n",
    "    exon_intron_data = pd.DataFrame({\n",
    "        \"Type\": [\"Exon\"] * len(exons) + [\"Intron\"] * len(introns),\n",
    "        \"Sequence\": exons + introns\n",
    "    })\n",
    "\n",
    "\n",
    "    print(f\"Exon and Intron Sequences {exon_intron_data}\")\n",
    "\n",
    "else:\n",
    "    print(f\"No sequence data found for {target_transcript}.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a334537-7923-48ac-bd66-1ff5e7d2219b",
   "metadata": {},
   "source": [
    "## check for shuffled nucelotide exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6de52751-e45a-4f7c-bb73-33ed293e3781",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "de bugging Adjusted Junctions for chr1:901862-911245\n",
      "Adjusted Exon Starts: [132, 321, 3941, 4119, 4276, 4524, 4726, 4922, 5668, 5942, 6528, 6844, 7158, 7569, 7882]\n",
      "Adjusted Exon Ends: [222, 3795, 4039, 4204, 4397, 4631, 4842, 5593, 5806, 6379, 6704, 7018, 7351, 7834, 7960]\n",
      "Sequence Length: 9384\n",
      "Exon and Intron Sequences       Type                                           Sequence\n",
      "0     Exon  TTAGTATGATCACATGGTGTGTAGAAGTTTGTTTTATAATTTTAAA...\n",
      "1     Exon  CGACGGGCGTGGTTCGTGAAGATCCTAAGTTACCATGGCTACTTTG...\n",
      "2     Exon  ATAACTGCGTGCAAGATCTTTACTGTATTAGTAGGGCGTACGTGAT...\n",
      "3     Exon  GCCCTAATATTCTTTAGCTCATTCCGGAATCAATTTAACTGGAGGT...\n",
      "4     Exon  CAATACTACTGATCTATAGAATAAAAGCTACGTTATTGCTCCGGTA...\n",
      "5     Exon  TACCGTTGAATTCAAAAAGGTTTTAGACTATTCGAAGGCAGATAGG...\n",
      "6     Exon  GAATACGATAGGGCAAGGATCCGTTTATCGAAGTGTTCATCATACG...\n",
      "7     Exon  GTTCGCTTAACGTAGACGAATATAGCTTTATCCCCCTGGAACGAAC...\n",
      "8     Exon  TGTACGGACTATTAGATTAACTACTAGGAGATAGCGTAGCAATAGT...\n",
      "9     Exon  TCTATGTATTATTACGCTTTTTAAAAGTGTATCGTGCCGATTCAAG...\n",
      "10    Exon  TGAGTTATACGGTCGTCAATATCGATTCCTTAAGATCTAGCATTTG...\n",
      "11    Exon  TCTGCAAATTACTCAAGATTATTAGTGTGGTTACTGCACTCCGCGC...\n",
      "12    Exon  GGCCTGTTATATTATCCCACGAGTTCCTGGTATTAAATAGTTTTGT...\n",
      "13    Exon  GATGGTCCGCTCCTATGAGTCTTATCACAGATATATCAAACAAGAG...\n",
      "14    Exon  ATTCAGTATCGCAAGCAAAGGTTCGGATTTGGTTTTCCAGGGGGAA...\n",
      "15  Intron  AGGAGGCTGTGGACAGGGACCCAGACTTGCCGACCTGTACGACTCT...\n",
      "16  Intron  AGAGGACAGCGCGCGGATGTCGGCCGGCCTGCCGGGCCCCGAGGCT...\n",
      "17  Intron  GACATCCTGGACCTGGAGAACCAGCGAGAAAACCTGGAGCAGCCAT...\n",
      "18  Intron  GATGTCAGCGACTGCTACCTGGAGCTATTCCCCGCCCACCTGTACT...\n",
      "19  Intron  GGGCTGTTACCGCTGACGGAGCTGAGTGTCTGCCCGCTCGAGGGGT...\n",
      "20  Intron  GCCCACTGCCCGCACCCCTCCTGGTGCTCTGCCCCAGCCGGGCCGA...\n",
      "21  Intron  CGCCGTCTAACCCGGCTGCGGACGGCGTCAGGGCACGAACCCGGCG...\n",
      "22  Intron  GAGCAGTGGGACCGGCTCTTGGTCCTGTACCCAACGTCCTTGGCCA...\n",
      "23  Intron  GGGGAGCTCCCACTCCGTGCCGTCCACATCAACCTGGAGGAGAAGG...\n",
      "24  Intron  GCCCCCTCATCAACACCATCCGCGTGGTGTGCGCCAGCTACGAGGA...\n",
      "25  Intron  GTTATGGGCAGTGGCCGAGGCTCACTCTCCTCAGGCGGACAGACCA...\n",
      "26  Intron  GACCAGGCCAACTCTGACCGTGCCAGCATTGGCCGACGGAGGACCG...\n",
      "27  Intron  CTGCACAGGCTGAGCCTGGAGAGCAGCCCAGATGCCCCTGACCACA...\n",
      "28  Intron  TTCCTCAGTGCCATGCAGAGTGCACGTGGACCCACGCCCTCGAGCC...\n",
      "29  Intron   GTCTCCTCTGCCAGGGAAGGTTCGCCCGAACCCTGGCTGCCTCTGACA\n",
      "30  Intron  ATGGTCGGTCCCCCAGGAGGAGCCGGGACCCCGGCTACGACCACCT...\n"
     ]
    }
   ],
   "source": [
    "shuffled_codons_file = \"/mnt/lareaulab/sdahiyat/illumina/shuffled_nt_new.txt\"\n",
    "canonical_dataset_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "\n",
    "sequences = {}\n",
    "with open(shuffled_codons_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue  \n",
    "        try:\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "        except ValueError:\n",
    "            print(f\"Skipping malformed line: {line}\")\n",
    "\n",
    "junctions = {}\n",
    "with open(canonical_dataset_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if len(fields) < 8: \n",
    "            continue\n",
    "\n",
    "        chrom = fields[2]  \n",
    "        start = int(fields[4]) \n",
    "        end = int(fields[5])\n",
    "        exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "        exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "\n",
    "        transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "        junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "\n",
    "if target_transcript in sequences and target_transcript in junctions:\n",
    "    full_sequence = sequences[target_transcript]  \n",
    "    exon_starts, exon_ends = junctions[target_transcript]  \n",
    "\n",
    "    seq_start = int(target_transcript.split(\":\")[1].split(\"-\")[0])  \n",
    "    exon_starts = [s - seq_start for s in exon_starts]\n",
    "    exon_ends = [e - seq_start for e in exon_ends]\n",
    "\n",
    "    print(f\"de bugging Adjusted Junctions for {target_transcript}\")\n",
    "    print(f\"Adjusted Exon Starts: {exon_starts}\")\n",
    "    print(f\"Adjusted Exon Ends: {exon_ends}\")\n",
    "    print(f\"Sequence Length: {len(full_sequence)}\")\n",
    "\n",
    "    exons = []\n",
    "    introns = []\n",
    "    last_pos = 0 \n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        if 0 <= exon_start < len(full_sequence) and 0 <= exon_end <= len(full_sequence):\n",
    "            introns.append(full_sequence[last_pos:exon_start])  \n",
    "            exons.append(full_sequence[exon_start:exon_end])  \n",
    "        else:\n",
    "            print(f\"Warning: Adjusted Exon range ({exon_start}-{exon_end}) out of bounds.\")\n",
    "        \n",
    "        last_pos = exon_end\n",
    "\n",
    "    introns.append(full_sequence[last_pos:])  \n",
    "\n",
    "\n",
    "    import pandas as pd\n",
    "    exon_intron_data = pd.DataFrame({\n",
    "        \"Type\": [\"Exon\"] * len(exons) + [\"Intron\"] * len(introns),\n",
    "        \"Sequence\": exons + introns\n",
    "    })\n",
    "\n",
    "\n",
    "    print(f\"Exon and Intron Sequences {exon_intron_data}\")\n",
    "\n",
    "else:\n",
    "    print(f\"No sequence data found for {target_transcript}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "00ca2a62-cfdf-4185-bdb3-36a473c6ca18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debugging Adjusted Junctions for chr1:901862-911245\n",
      "Adjusted Exon Starts: [132, 321, 3941, 4119, 4276, 4524, 4726, 4922, 5668, 5942, 6528, 6844, 7158, 7569, 7882]\n",
      "Adjusted Exon Ends: [222, 3795, 4039, 4204, 4397, 4631, 4842, 5593, 5806, 6379, 6704, 7018, 7351, 7834, 7960]\n",
      "Sequence Length: 9384\n",
      "Exon and Intron Sequences       Type                                           Sequence\n",
      "0     Exon  TCACTAAAGCACCAGACGCAGGTGTCATTGAAATTCCATATCTTGA...\n",
      "1     Exon  TACCACGATATATCAAAGTTGTTGTCTCAAGTTGATTGGAAGAAAC...\n",
      "2     Exon  ACCAAGGAAAACACTCACGTGCTCAAAAACGAAAAATGAACAACCG...\n",
      "3     Exon  CAGCCACAGGACTTTTGTTTGTGTTGGTTCCACAAGCTGATCGCAT...\n",
      "4     Exon  GTTCACCAGACAAGGTGGCTATTACTGTTTGGTATCAGGATGAGTG...\n",
      "5     Exon  AGGCCAAGCTCTTGATCAATATGAAGGGAAGGTCTGACCCTTCTGC...\n",
      "6     Exon  TACTATCCCCTCCCTCTCAGTTGTGTAGTATCGAGTCTTTGATAAA...\n",
      "7     Exon  TTTATTAGAGGAAGTAAACAACGAACCAGATAAAATTTTGCTAAAA...\n",
      "8     Exon  CACGGTGCTTGGTGAAGGACAAAATCCCTGTTGCCTGGAATGTTAA...\n",
      "9     Exon  TAAATGACACTTCCCAGCACATGGACTTGTAGAACGGATCAAGTAA...\n",
      "10    Exon  CTCCACTCCACTCTATCTTTGAACCCCATTGACAGGAATTCAATTA...\n",
      "11    Exon  TGAAATAGGAATCGGGGTAAACTTCATTTCCCAGCAATATCCTTAA...\n",
      "12    Exon  CTTCGTACCAGCGGGATCAGAGCGTAGATTAATTGGCGTTTCGCTG...\n",
      "13    Exon  ACTGGTCCCTAAATTCTCCACTAGCAGGGTATTAACCTGTGGTCCA...\n",
      "14    Exon  CTTAAATAATTGCTTGCTTATCGAAAATTGCAGTGATTCCTGATCG...\n",
      "15  Intron  AGGAGGCTGTGGACAGGGACCCAGACTTGCCGACCTGTACGACTCT...\n",
      "16  Intron  AGAGGACAGCGCGCGGATGTCGGCCGGCCTGCCGGGCCCCGAGGCT...\n",
      "17  Intron  GACATCCTGGACCTGGAGAACCAGCGAGAAAACCTGGAGCAGCCAT...\n",
      "18  Intron  GATGTCAGCGACTGCTACCTGGAGCTATTCCCCGCCCACCTGTACT...\n",
      "19  Intron  GGGCTGTTACCGCTGACGGAGCTGAGTGTCTGCCCGCTCGAGGGGT...\n",
      "20  Intron  GCCCACTGCCCGCACCCCTCCTGGTGCTCTGCCCCAGCCGGGCCGA...\n",
      "21  Intron  CGCCGTCTAACCCGGCTGCGGACGGCGTCAGGGCACGAACCCGGCG...\n",
      "22  Intron  GAGCAGTGGGACCGGCTCTTGGTCCTGTACCCAACGTCCTTGGCCA...\n",
      "23  Intron  GGGGAGCTCCCACTCCGTGCCGTCCACATCAACCTGGAGGAGAAGG...\n",
      "24  Intron  GCCCCCTCATCAACACCATCCGCGTGGTGTGCGCCAGCTACGAGGA...\n",
      "25  Intron  GTTATGGGCAGTGGCCGAGGCTCACTCTCCTCAGGCGGACAGACCA...\n",
      "26  Intron  GACCAGGCCAACTCTGACCGTGCCAGCATTGGCCGACGGAGGACCG...\n",
      "27  Intron  CTGCACAGGCTGAGCCTGGAGAGCAGCCCAGATGCCCCTGACCACA...\n",
      "28  Intron  TTCCTCAGTGCCATGCAGAGTGCACGTGGACCCACGCCCTCGAGCC...\n",
      "29  Intron   GTCTCCTCTGCCAGGGAAGGTTCGCCCGAACCCTGGCTGCCTCTGACA\n",
      "30  Intron  ATGGTCGGTCCCCCAGGAGGAGCCGGGACCCCGGCTACGACCACCT...\n"
     ]
    }
   ],
   "source": [
    "shuffled_codons_file = \"/mnt/lareaulab/sdahiyat/illumina/shuffled_codons_from_codon_freq.txt\"\n",
    "canonical_dataset_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "# define the target transcript\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "\n",
    "sequences = {}\n",
    "with open(shuffled_codons_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue  # Skip empty lines\n",
    "        try:\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "        except ValueError:\n",
    "            print(f\"Skipping malformed line: {line}\")\n",
    "\n",
    "junctions = {}\n",
    "with open(canonical_dataset_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if len(fields) < 8:  \n",
    "            continue\n",
    "\n",
    "        chrom = fields[2]  # Chromosome\n",
    "        start = int(fields[4])  # Start position\n",
    "        end = int(fields[5])    # End position\n",
    "        exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "        exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "\n",
    "        transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "        junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "\n",
    "if target_transcript in sequences and target_transcript in junctions:\n",
    "    full_sequence = sequences[target_transcript] \n",
    "    exon_starts, exon_ends = junctions[target_transcript]  \n",
    "\n",
    "    seq_start = int(target_transcript.split(\":\")[1].split(\"-\")[0])  \n",
    "    exon_starts = [s - seq_start for s in exon_starts]\n",
    "    exon_ends = [e - seq_start for e in exon_ends]\n",
    "\n",
    "    print(f\"Debugging Adjusted Junctions for {target_transcript}\")\n",
    "    print(f\"Adjusted Exon Starts: {exon_starts}\")\n",
    "    print(f\"Adjusted Exon Ends: {exon_ends}\")\n",
    "    print(f\"Sequence Length: {len(full_sequence)}\")\n",
    "\n",
    "    exons = []\n",
    "    introns = []\n",
    "    last_pos = 0 \n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        if 0 <= exon_start < len(full_sequence) and 0 <= exon_end <= len(full_sequence):\n",
    "            introns.append(full_sequence[last_pos:exon_start])  \n",
    "            exons.append(full_sequence[exon_start:exon_end])  \n",
    "        else:\n",
    "            print(f\"Warning: Adjusted Exon range ({exon_start}-{exon_end}) out of bounds.\")\n",
    "        \n",
    "        last_pos = exon_end\n",
    "\n",
    "    introns.append(full_sequence[last_pos:])  \n",
    "\n",
    "    import pandas as pd\n",
    "    exon_intron_data = pd.DataFrame({\n",
    "        \"Type\": [\"Exon\"] * len(exons) + [\"Intron\"] * len(introns),\n",
    "        \"Sequence\": exons + introns\n",
    "    })\n",
    "\n",
    "    \n",
    "\n",
    "    print(f\"Exon and Intron Sequences {exon_intron_data}\")\n",
    "\n",
    "else:\n",
    "    print(f\"No sequence data found for {target_transcript}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "36330669-4d1c-4142-8100-927494da1b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debugging Adjusted Junctions for chr1:901862-911245\n",
      "Adjusted Exon Starts: [132, 321, 3941, 4119, 4276, 4524, 4726, 4922, 5668, 5942, 6528, 6844, 7158, 7569, 7882]\n",
      "Adjusted Exon Ends: [222, 3795, 4039, 4204, 4397, 4631, 4842, 5593, 5806, 6379, 6704, 7018, 7351, 7834, 7960]\n",
      "Sequence Length: 9384\n",
      "Exon and Intron Sequences       Type                                           Sequence\n",
      "0     Exon  GGTGAGCGGGGCGTGGGTGCGGCCACCTGGGCGCAGGGCTCCCCCA...\n",
      "1     Exon  GGTGAGCGCGGCGTGCACGGTGGCTGTGGTCTGGGAGCGTGGCTCT...\n",
      "2     Exon  GGTGGGGGCCGGGCTGGGTGGAGCACGCTAAGGGTGCAGCATCCCC...\n",
      "3     Exon  GGTGAGGCGGTGGGCAATGGGGTGGGGCCATGGCCGCCCTTCCCTC...\n",
      "4     Exon  GGTGTTTGGGATGCTTCCCGGGCCCCCAGAGGCACTCCTGACCCAG...\n",
      "5     Exon  GGTCAGTGCCGGGGACCCCACCCCCCTCCCCACCCTGATCCTCGCA...\n",
      "6     Exon  GGTGGGTGGGAGGTGCGTGGGGCTGTAGGGGGATGGGAGGGGTGCA...\n",
      "7     Exon  GGTGGGCCCCTCCCCACTGTGGGCCCGCCCCAGGGAGGCAGCTGTG...\n",
      "8     Exon  GGTAGGGCCCTGACCCTGGTTCTGCCTCCCGCCTGGCCAGGCCATG...\n",
      "9     Exon  GGTGAGGGGTCAATAGGCCCCACAGCCCAGGTCCTGGGCAGTGGTA...\n",
      "10    Exon  GGTGAGCGCTTACGGGGTGGCAGACGAAAGTGGGGCAGAAGGCTGT...\n",
      "11    Exon  GGTGGGCCCAGCACACCCACACAGCCCCTGGCCTGGTTCCCACCGT...\n",
      "12    Exon  GGTCAGGCCCCTGCTGGGTGACAGAAAGGGTGGGAGGTGCCTGCAG...\n",
      "13    Exon  TGTGAGTAGCAGCCCCCACGCCCGTGTGCCCCGGGCTCCGGGCTGG...\n",
      "14    Exon  GGTGAGTAAGGATCCTGCCTCCTGAGGTGAGTGCCTGTTGCCTCCC...\n",
      "15  Intron  AGGAGGCTGTGGACAGGGACCCAGACTTGCCGACCTGTACGACTCT...\n",
      "16  Intron  AGAGGACAGCGCGCGGATGTCGGCCGGCCTGCCGGGCCCCGAGGCT...\n",
      "17  Intron  GACATCCTGGACCTGGAGAACCAGCGAGAAAACCTGGAGCAGCCAT...\n",
      "18  Intron  GATGTCAGCGACTGCTACCTGGAGCTATTCCCCGCCCACCTGTACT...\n",
      "19  Intron  GGGCTGTTACCGCTGACGGAGCTGAGTGTCTGCCCGCTCGAGGGGT...\n",
      "20  Intron  GCCCACTGCCCGCACCCCTCCTGGTGCTCTGCCCCAGCCGGGCCGA...\n",
      "21  Intron  CGCCGTCTAACCCGGCTGCGGACGGCGTCAGGGCACGAACCCGGCG...\n",
      "22  Intron  GAGCAGTGGGACCGGCTCTTGGTCCTGTACCCAACGTCCTTGGCCA...\n",
      "23  Intron  GGGGAGCTCCCACTCCGTGCCGTCCACATCAACCTGGAGGAGAAGG...\n",
      "24  Intron  GCCCCCTCATCAACACCATCCGCGTGGTGTGCGCCAGCTACGAGGA...\n",
      "25  Intron  GTTATGGGCAGTGGCCGAGGCTCACTCTCCTCAGGCGGACAGACCA...\n",
      "26  Intron  GACCAGGCCAACTCTGACCGTGCCAGCATTGGCGACGGCAGGACCG...\n",
      "27  Intron  CTGCACAGGCTGAGCCTGGAGAGCAGCCCAGATGCCCCTGACCACA...\n",
      "28  Intron  TTCCTCAGTGCCATGCAGAGTGCACGTGGACCCACGCCCTCGAGCC...\n",
      "29  Intron   GTCTCCTCTGCCAGGGAAGGTTCGCCCGAACCCTGGCTGCCTCTGACA\n",
      "30  Intron  ATGGTCGGTCCCCCAGGAGGAGCCGGGACCCCGGCTACGACCACCT...\n"
     ]
    }
   ],
   "source": [
    "shuffled_codons_file = \"/mnt/lareaulab/sdahiyat/datasets/srsf1_matches_shuffled.txt\"\n",
    "canonical_dataset_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_filtered_motifs.txt\"\n",
    "\n",
    "# define the target transcript\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "\n",
    "sequences = {}\n",
    "with open(shuffled_codons_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue  # Skip empty lines\n",
    "        try:\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "        except ValueError:\n",
    "            print(f\"Skipping malformed line: {line}\")\n",
    "\n",
    "junctions = {}\n",
    "with open(canonical_dataset_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if len(fields) < 8:  \n",
    "            continue\n",
    "\n",
    "        chrom = fields[2]  # Chromosome\n",
    "        start = int(fields[4])  # Start position\n",
    "        end = int(fields[5])    # End position\n",
    "        exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "        exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "\n",
    "        transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "        junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "\n",
    "if target_transcript in sequences and target_transcript in junctions:\n",
    "    full_sequence = sequences[target_transcript] \n",
    "    exon_starts, exon_ends = junctions[target_transcript]  \n",
    "\n",
    "    seq_start = int(target_transcript.split(\":\")[1].split(\"-\")[0])  \n",
    "    exon_starts = [s - seq_start for s in exon_starts]\n",
    "    exon_ends = [e - seq_start for e in exon_ends]\n",
    "\n",
    "    print(f\"Debugging Adjusted Junctions for {target_transcript}\")\n",
    "    print(f\"Adjusted Exon Starts: {exon_starts}\")\n",
    "    print(f\"Adjusted Exon Ends: {exon_ends}\")\n",
    "    print(f\"Sequence Length: {len(full_sequence)}\")\n",
    "\n",
    "    exons = []\n",
    "    introns = []\n",
    "    last_pos = 0 \n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        if 0 <= exon_start < len(full_sequence) and 0 <= exon_end <= len(full_sequence):\n",
    "            introns.append(full_sequence[last_pos:exon_start])  \n",
    "            exons.append(full_sequence[exon_start:exon_end])  \n",
    "        else:\n",
    "            print(f\"Warning: Adjusted Exon range ({exon_start}-{exon_end}) out of bounds.\")\n",
    "        \n",
    "        last_pos = exon_end\n",
    "\n",
    "    introns.append(full_sequence[last_pos:])  \n",
    "\n",
    "    import pandas as pd\n",
    "    exon_intron_data = pd.DataFrame({\n",
    "        \"Type\": [\"Exon\"] * len(exons) + [\"Intron\"] * len(introns),\n",
    "        \"Sequence\": exons + introns\n",
    "    })\n",
    "\n",
    "\n",
    "    print(f\"Exon and Intron Sequences {exon_intron_data}\")\n",
    "\n",
    "else:\n",
    "    print(f\"No sequence data found for {target_transcript}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bfb7856c-2ce7-4ba8-a67f-ceb4d418aab4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debugging Adjusted Junctions for chr1:901862-911245\n",
      "Adjusted Exon Starts: [132, 321, 3941, 4119, 4276, 4524, 4726, 4922, 5668, 5942, 6528, 6844, 7158, 7569, 7882]\n",
      "Adjusted Exon Ends: [222, 3795, 4039, 4204, 4397, 4631, 4842, 5593, 5806, 6379, 6704, 7018, 7351, 7834, 7960]\n",
      "Sequence Length: 9384\n",
      "Exon and Intron Sequences       Type                                           Sequence\n",
      "0     Exon  AAGCCGGGGGGTCCTCAGCCTATTAAGGGCCGCCGTCCAGGGTCGG...\n",
      "1     Exon  tGtGGAtgaCgCTTCCTacggaGtgcaGCtAGGtCATGgtCcGCGT...\n",
      "2     Exon  CCCGGAGGCTCTGGCCTCATTTCCGGACACCGCCCTCTGGGGGGAG...\n",
      "3     Exon  CCGCAGTGGCCGGGCCCTTACATAGTCCCCCCAGGCGGAGGGGCCG...\n",
      "4     Exon  CCGCTGCGCGATGGGCCGGGGCGCCTTCCGCTGTAGTACGGCCGTC...\n",
      "5     Exon  CCTGCCCAGTAAGAACTACCGGTGGTACGGCCATCGCTCCCCCCGC...\n",
      "6     Exon  GTGGTCAGGTCCGCGGAGCCTTGCGCGGGTGGCGCCCGCACCCGAG...\n",
      "7     Exon  GCGCGACAGCTTCGCACGTTGGGGCCGCTTGAGTTGCCCCGGTGGG...\n",
      "8     Exon  GGGCGTATCGTGTGTCACTTGGGCCAGACACGCAGCGTCCTCCGAC...\n",
      "9     Exon  CggatGcgagcacCGagtaacaGagAAGAaGagcaactagCACgaC...\n",
      "10    Exon  CCTAAACAGAGGATTAGGGGCGGACGGTCGGCTCGTAGAAGCTAGT...\n",
      "11    Exon  GTCCTACGCCCTCCCGCGCACACCGACTGGTGTGACTCCTTCCTCC...\n",
      "12    Exon  GGCGACCCGTGCAGCCACATCGGGGCGTTGGAGATCCGGCAGCGTA...\n",
      "13    Exon  GGGTCCGGCTTCGGGAGGCAGCGGGCGTGACTGGCCAGCTTGCCAC...\n",
      "14    Exon  TCACGTCCAGCCGCGGGGCTTACTTCGTCCCGAGAACTCGCTTGGA...\n",
      "15  Intron  AGGAGGCTGTGGACAGGGACCCAGACTTGCCGACCTGTACGACTCT...\n",
      "16  Intron  AGAGGACAGCGCGCGGATGTCGGCCGGCCTGCCGGGCCCCGAGGCT...\n",
      "17  Intron  GACATCCTGGACCTGGAGAACCAGCGAGAAAACCTGGAGCAGCCAT...\n",
      "18  Intron  GATGTCAGCGACTGCTACCTGGAGCTATTCCCCGCCCACCTGTACT...\n",
      "19  Intron  GGGCTGTTACCGCTGACGGAGCTGAGTGTCTGCCCGCTCGAGGGGT...\n",
      "20  Intron  GCCCACTGCCCGCACCCCTCCTGGTGCTCTGCCCCAGCCGGGCCGA...\n",
      "21  Intron  CGCCGTCTAACCCGGCTGCGGACGGCGTCAGGGCACGAACCCGGCG...\n",
      "22  Intron  GAGCAGTGGGACCGGCTCTTGGTCCTGTACCCAACGTCCTTGGCCA...\n",
      "23  Intron  GGGGAGCTCCCACTCCGTGCCGTCCACATCAACCTGGAGGAGAAGG...\n",
      "24  Intron  GCCCCCTCATCAACACCATCCGCGTGGTGTGCGCCAGCTACGAGGA...\n",
      "25  Intron  GTTATGGGCAGTGGCCGAGGCTCACTCTCCTCAGGCGGACAGACCA...\n",
      "26  Intron  GACCAGGCCAACTCTGACCGTGCCAGCATTGGCCGACGGAGGACCG...\n",
      "27  Intron  CTGCACAGGCTGAGCCTGGAGAGCAGCCCAGATGCCCCTGACCACA...\n",
      "28  Intron  TTCCTCAGTGCCATGCAGAGTGCACGTGGACCCACGCCCTCGAGCC...\n",
      "29  Intron   GTCTCCTCTGCCAGGGAAGGTTCGCCCGAACCCTGGCTGCCTCTGACA\n",
      "30  Intron  ATGGTCGGTCCCCCAGGAGGAGCCGGGACCCCGGCTACGACCACCT...\n"
     ]
    }
   ],
   "source": [
    "shuffled_codons_file = \"/mnt/lareaulab/sdahiyat/illumina/permuted_exon_sequences.txt\"\n",
    "canonical_dataset_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "\n",
    "sequences = {}\n",
    "with open(shuffled_codons_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue  \n",
    "        try:\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "        except ValueError:\n",
    "            print(f\"Skipping malformed line: {line}\")\n",
    "\n",
    "junctions = {}\n",
    "with open(canonical_dataset_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if len(fields) < 8:  \n",
    "            continue\n",
    "\n",
    "        chrom = fields[2]  # Chromosome\n",
    "        start = int(fields[4])  # Start position\n",
    "        end = int(fields[5])    # End position\n",
    "        exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "        exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "\n",
    "        transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "        junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "\n",
    "if target_transcript in sequences and target_transcript in junctions:\n",
    "    full_sequence = sequences[target_transcript] \n",
    "    exon_starts, exon_ends = junctions[target_transcript]  \n",
    "\n",
    "    seq_start = int(target_transcript.split(\":\")[1].split(\"-\")[0])  \n",
    "    exon_starts = [s - seq_start for s in exon_starts]\n",
    "    exon_ends = [e - seq_start for e in exon_ends]\n",
    "\n",
    "    print(f\"Debugging Adjusted Junctions for {target_transcript}\")\n",
    "    print(f\"Adjusted Exon Starts: {exon_starts}\")\n",
    "    print(f\"Adjusted Exon Ends: {exon_ends}\")\n",
    "    print(f\"Sequence Length: {len(full_sequence)}\")\n",
    "\n",
    "    exons = []\n",
    "    introns = []\n",
    "    last_pos = 0 \n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        if 0 <= exon_start < len(full_sequence) and 0 <= exon_end <= len(full_sequence):\n",
    "            introns.append(full_sequence[last_pos:exon_start])  \n",
    "            exons.append(full_sequence[exon_start:exon_end])  \n",
    "        else:\n",
    "            print(f\"Warning: Adjusted Exon range ({exon_start}-{exon_end}) out of bounds.\")\n",
    "        \n",
    "        last_pos = exon_end\n",
    "\n",
    "    introns.append(full_sequence[last_pos:])  \n",
    "\n",
    "    import pandas as pd\n",
    "    exon_intron_data = pd.DataFrame({\n",
    "        \"Type\": [\"Exon\"] * len(exons) + [\"Intron\"] * len(introns),\n",
    "        \"Sequence\": exons + introns\n",
    "    })\n",
    "\n",
    "\n",
    "    print(f\"Exon and Intron Sequences {exon_intron_data}\")\n",
    "\n",
    "else:\n",
    "    print(f\"No sequence data found for {target_transcript}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1143d632-0343-42da-8df0-e70ee7d9595f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ Comparing exons for transcript: chr1:901862-911245\n",
      "\n",
      "Exon 1 [901994-902084]\n",
      "Original : GGTGAGCGGGGCGTGGGTGCGGCCACCTGGGCGCAGGGCTCCCCCACCCGCTCCGGGGCC\n",
      "Shuffled : AAGCCGGGGGGTCCTCAGCCTATTAAGGGCCGCCGTCCAGGGTCGGCGCCGTGCTCCCGC\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 2 [902183-905657]\n",
      "Original : GGTGAGCGCGGCGTGCACGGTGGCTGTGGTCTGGGAGCGTGGCTCTGCCCGCGCGTGTGT\n",
      "Shuffled : tGtGGAtgaCgCTTCCTacggaGtgcaGCtAGGtCATGgtCcGCGTccTTGTaCGGGCAg\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 3 [905803-905901]\n",
      "Original : GGTGGGGGCCGGGCTGGGTGGAGCACGCTAAGGGTGCAGCATCCCCATCAGCCTGGGGCT\n",
      "Shuffled : CCCGGAGGCTCTGGCCTCATTTCCGGACACCGCCCTCTGGGGGGAGGGCTGAGTGGTGCA\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 4 [905981-906066]\n",
      "Original : GGTGAGGCGGTGGGCAATGGGGTGGGGCCATGGCCGCCCTTCCCTCCATGGATCCCTGAA\n",
      "Shuffled : CCGCAGTGGCCGGGCCCTTACATAGTCCCCCCAGGCGGAGGGGCCGTGTGTGTCCGCAGT\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 5 [906138-906259]\n",
      "Original : GGTGTTTGGGATGCTTCCCGGGCCCCCAGAGGCACTCCTGACCCAGGACTTGGAGAGGGG\n",
      "Shuffled : CCGCTGCGCGATGGGCCGGGGCGCCTTCCGCTGTAGTACGGCCGTCGATGAGAGTTCGTG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 6 [906386-906493]\n",
      "Original : GGTCAGTGCCGGGGACCCCACCCCCCTCCCCACCCTGATCCTCGCAGCCGGCTCTGACCT\n",
      "Shuffled : CCTGCCCAGTAAGAACTACCGGTGGTACGGCCATCGCTCCCCCCGCCACAGACTCGCCTC\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 7 [906588-906704]\n",
      "Original : GGTGGGTGGGAGGTGCGTGGGGCTGTAGGGGGATGGGAGGGGTGCATGGTGGTGGGCAGG\n",
      "Shuffled : GTGGTCAGGTCCGCGGAGCCTTGCGCGGGTGGCGCCCGCACCCGAGATTAGGGGGAGCTT\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 8 [906784-907455]\n",
      "Original : GGTGGGCCCCTCCCCACTGTGGGCCCGCCCCAGGGAGGCAGCTGTGTGGGGGTGGGAGGG\n",
      "Shuffled : GCGCGACAGCTTCGCACGTTGGGGCCGCTTGAGTTGCCCCGGTGGGTCCCACCGTATGTT\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 9 [907530-907668]\n",
      "Original : GGTAGGGCCCTGACCCTGGTTCTGCCTCCCGCCTGGCCAGGCCATGGTGGGGCGGGAGCC\n",
      "Shuffled : GGGCGTATCGTGTGTCACTTGGGCCAGACACGCAGCGTCCTCCGACGGGAGCTGGCCCAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 10 [907804-908241]\n",
      "Original : GGTGAGGGGTCAATAGGCCCCACAGCCCAGGTCCTGGGCAGTGGTAAAAAGGGGGCAGCA\n",
      "Shuffled : CggatGcgagcacCGagtaacaGagAAGAaGagcaactagCACgaCggGcgTgtGaGTCc\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 11 [908390-908566]\n",
      "Original : GGTGAGCGCTTACGGGGTGGCAGACGAAAGTGGGGCAGAAGGCTGTCGGGTAGGTGTGTG\n",
      "Shuffled : CCTAAACAGAGGATTAGGGGCGGACGGTCGGCTCGTAGAAGCTAGTCACAGGAGTGATCC\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 12 [908706-908880]\n",
      "Original : GGTGGGCCCAGCACACCCACACAGCCCCTGGCCTGGTTCCCACCGTTCCGCACCACTGGT\n",
      "Shuffled : GTCCTACGCCCTCCCGCGCACACCGACTGGTGTGACTCCTTCCTCCGTGATGCTTGTCTG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 13 [909020-909213]\n",
      "Original : GGTCAGGCCCCTGCTGGGTGACAGAAAGGGTGGGAGGTGCCTGCAGCCTGAGGCTGGGGA\n",
      "Shuffled : GGCGACCCGTGCAGCCACATCGGGGCGTTGGAGATCCGGCAGCGTACGTCTGGGACCTCT\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 14 [909431-909696]\n",
      "Original : TGTGAGTAGCAGCCCCCACGCCCGTGTGCCCCGGGCTCCGGGCTGGCCGGGGGTCTGGTG\n",
      "Shuffled : GGGTCCGGCTTCGGGAGGCAGCGGGCGTGACTGGCCAGCTTGCCACGCGGGGCGCCGTCG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 15 [909744-909822]\n",
      "Original : GGTGAGTAAGGATCCTGCCTCCTGAGGTGAGTGCCTGTTGCCTCCCACAGGCTGACACAT\n",
      "Shuffled : TCACGTCCAGCCGCGGGGCTTACTTCGTCCCGAGAACTCGCTTGGACGTGTAATTTCAAC\n",
      "✅ DIFFERENT — exon was shuffled\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "canonical_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_sequence_unflanked_unshuffled.txt\"\n",
    "shuffled_file = \"/mnt/lareaulab/sdahiyat/illumina/permuted_exon_sequences.txt\"\n",
    "junction_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "def load_sequences(file_path):\n",
    "    sequences = {}\n",
    "    with open(file_path, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if not line or \"\\t\" not in line:\n",
    "                continue\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "    return sequences\n",
    "\n",
    "original_sequences = load_sequences(canonical_file)\n",
    "shuffled_sequences = load_sequences(shuffled_file)\n",
    "\n",
    "def load_junctions(file_path):\n",
    "    junctions = {}\n",
    "    with open(file_path, \"r\") as f:\n",
    "        for line in f:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if len(fields) < 8:\n",
    "                continue\n",
    "            chrom = fields[2]\n",
    "            start = int(fields[4])\n",
    "            end = int(fields[5])\n",
    "            exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "            exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "            transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "            junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "    return junctions\n",
    "\n",
    "junctions = load_junctions(junction_file)\n",
    "\n",
    "def compare_original_and_shuffled_exons(transcript_key, canonical, shuffled, junctions):\n",
    "    if transcript_key not in canonical or transcript_key not in shuffled:\n",
    "        print(\"Transcript not found in one of the sequence files.\")\n",
    "        return\n",
    "\n",
    "    if transcript_key not in junctions:\n",
    "        print(\"Transcript not found in junction file.\")\n",
    "        return\n",
    "\n",
    "    original_seq = canonical[transcript_key]\n",
    "    shuffled_seq = shuffled[transcript_key]\n",
    "    exon_starts, exon_ends = junctions[transcript_key]\n",
    "    seq_start = int(transcript_key.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "    print(f\"\\n Comparing exons for transcript: {transcript_key}\")\n",
    "    for i, (abs_start, abs_end) in enumerate(zip(exon_starts, exon_ends)):\n",
    "        start = abs_start - seq_start\n",
    "        end = abs_end - seq_start\n",
    "        if start < 0 or end > len(original_seq):\n",
    "            print(f\"⚠Skipping out-of-bounds exon {i+1}: {abs_start}-{abs_end}\")\n",
    "            continue\n",
    "\n",
    "        original_exon = original_seq[start:end]\n",
    "        shuffled_exon = shuffled_seq[start:end]\n",
    "\n",
    "        print(f\"\\nExon {i+1} [{abs_start}-{abs_end}]\")\n",
    "        print(f\"Original : {original_exon[:60]}\")\n",
    "        print(f\"Shuffled : {shuffled_exon[:60]}\")\n",
    "        if original_exon == shuffled_exon:\n",
    "            print(\"SAME — not shuffled or identical by chance\")\n",
    "        else:\n",
    "            print(\"DIFFERENT — exon was shuffled\")\n",
    "\n",
    "compare_original_and_shuffled_exons(\n",
    "    transcript_key=target_transcript,\n",
    "    canonical=original_sequences,\n",
    "    shuffled=shuffled_sequences,\n",
    "    junctions=junctions\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "21f0fb3e-8664-4241-9b63-544876e02110",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debugging Adjusted Junctions for chr1:901862-911245\n",
      "Adjusted Exon Starts: [132, 321, 3941, 4119, 4276, 4524, 4726, 4922, 5668, 5942, 6528, 6844, 7158, 7569, 7882]\n",
      "Adjusted Exon Ends: [222, 3795, 4039, 4204, 4397, 4631, 4842, 5593, 5806, 6379, 6704, 7018, 7351, 7834, 7960]\n",
      "Sequence Length: 9384\n",
      "Exon and Intron Sequences       Type                                           Sequence\n",
      "0     Exon  GGCACCCCTCGCGGTCCCGACGCCTGGGCGGGCCCATCCCAGCGGC...\n",
      "1     Exon  attGGAcaaGAGtatgtgTGTCTCCTGGGGggtAGGCCATGGgcaT...\n",
      "2     Exon  GGCCAGCATCCGCTCTGTCCGAGCCAGGCTGGGCGACGCCCGCCCT...\n",
      "3     Exon  GGCCTGGCTCTCCGGATGGCTAAGGGTGCCCGCAGGGCTCTGTCAT...\n",
      "4     Exon  GTTTGCATGTGTGGACGTGGAGGGGCTGGGCGGGCTCCCGCCCCAG...\n",
      "5     Exon  GGCCGTCATCCTGCGGAGCTCCATCCTTGCAGGACGAAGGGCTGCC...\n",
      "6     Exon  GGTGGCCTAGGCATGCAGGACTGGGGCAGGCAGCTACCGGTGGGCC...\n",
      "7     Exon  GGCGTTCACTCGGGCTGGAGGGGGCTAGGGGACCTCCCGCTAGGGC...\n",
      "8     Exon  GGGCGCGCAGCGCAGAGGCCCGGTGTAGTTGACGGTCCCACGGCCG...\n",
      "9     Exon  GGtgttggCTGcggtgtaccGGAattgagCCAgctgcaATCaaact...\n",
      "10    Exon  GGCAGGGGGGGTGGGGTCAGCCAAAGTCCGGGAGAGTTGCGAGAAA...\n",
      "11    Exon  AGCCAGCAGGCCTCCTCTCCCACACCCTCTCCAACAGCCAAGCTCC...\n",
      "12    Exon  GCCCTGGGCCTGGCCAGACAGCGCTTCCGGGGTTGTGGGTCCACAG...\n",
      "13    Exon  TTGGCCCGCCGTCGGGAGATTGTGGCCTGTGCCGAGTGGACCGGGG...\n",
      "14    Exon  CACTGACTTCTGACATCCCTGGAGGGTTCCTGCAGGGCCCCCCAGG...\n",
      "15  Intron  AGGAGGCTGTGGACAGGGACCCAGACTTGCCGACCTGTACGACTCT...\n",
      "16  Intron  AGAGGACAGCGCGCGGATGTCGGCCGGCCTGCCGGGCCCCGAGGCT...\n",
      "17  Intron  GACATCCTGGACCTGGAGAACCAGCGAGAAAACCTGGAGCAGCCAT...\n",
      "18  Intron  GATGTCAGCGACTGCTACCTGGAGCTATTCCCCGCCCACCTGTACT...\n",
      "19  Intron  GGGCTGTTACCGCTGACGGAGCTGAGTGTCTGCCCGCTCGAGGGGT...\n",
      "20  Intron  GCCCACTGCCCGCACCCCTCCTGGTGCTCTGCCCCAGCCGGGCCGA...\n",
      "21  Intron  CGCCGTCTAACCCGGCTGCGGACGGCGTCAGGGCACGAACCCGGCG...\n",
      "22  Intron  GAGCAGTGGGACCGGCTCTTGGTCCTGTACCCAACGTCCTTGGCCA...\n",
      "23  Intron  GGGGAGCTCCCACTCCGTGCCGTCCACATCAACCTGGAGGAGAAGG...\n",
      "24  Intron  GCCCCCTCATCAACACCATCCGCGTGGTGTGCGCCAGCTACGAGGA...\n",
      "25  Intron  GTTATGGGCAGTGGCCGAGGCTCACTCTCCTCAGGCGGACAGACCA...\n",
      "26  Intron  GACCAGGCCAACTCTGACCGTGCCAGCATTGGCCGACGGAGGACCG...\n",
      "27  Intron  CTGCACAGGCTGAGCCTGGAGAGCAGCCCAGATGCCCCTGACCACA...\n",
      "28  Intron  TTCCTCAGTGCCATGCAGAGTGCACGTGGACCCACGCCCTCGAGCC...\n",
      "29  Intron   GTCTCCTCTGCCAGGGAAGGTTCGCCCGAACCCTGGCTGCCTCTGACA\n",
      "30  Intron  ATGGTCGGTCCCCCAGGAGGAGCCGGGACCCCGGCTACGACCACCT...\n"
     ]
    }
   ],
   "source": [
    "shuffled_codons_file = \"/mnt/lareaulab/sdahiyat/illumina/permuted_exons_by_codon.txt\"\n",
    "canonical_dataset_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "\n",
    "sequences = {}\n",
    "with open(shuffled_codons_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue  \n",
    "        try:\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "        except ValueError:\n",
    "            print(f\"Skipping malformed line: {line}\")\n",
    "\n",
    "junctions = {}\n",
    "with open(canonical_dataset_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if len(fields) < 8:  \n",
    "            continue\n",
    "\n",
    "        chrom = fields[2]  # Chromosome\n",
    "        start = int(fields[4])  # Start position\n",
    "        end = int(fields[5])    # End position\n",
    "        exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "        exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "\n",
    "        transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "        junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "\n",
    "if target_transcript in sequences and target_transcript in junctions:\n",
    "    full_sequence = sequences[target_transcript] \n",
    "    exon_starts, exon_ends = junctions[target_transcript]  \n",
    "\n",
    "    seq_start = int(target_transcript.split(\":\")[1].split(\"-\")[0])  \n",
    "    exon_starts = [s - seq_start for s in exon_starts]\n",
    "    exon_ends = [e - seq_start for e in exon_ends]\n",
    "\n",
    "    print(f\"Debugging Adjusted Junctions for {target_transcript}\")\n",
    "    print(f\"Adjusted Exon Starts: {exon_starts}\")\n",
    "    print(f\"Adjusted Exon Ends: {exon_ends}\")\n",
    "    print(f\"Sequence Length: {len(full_sequence)}\")\n",
    "\n",
    "    exons = []\n",
    "    introns = []\n",
    "    last_pos = 0 \n",
    "\n",
    "    for exon_start, exon_end in zip(exon_starts, exon_ends):\n",
    "        if 0 <= exon_start < len(full_sequence) and 0 <= exon_end <= len(full_sequence):\n",
    "            introns.append(full_sequence[last_pos:exon_start])  \n",
    "            exons.append(full_sequence[exon_start:exon_end])  \n",
    "        else:\n",
    "            print(f\"Warning: Adjusted Exon range ({exon_start}-{exon_end}) out of bounds.\")\n",
    "        \n",
    "        last_pos = exon_end\n",
    "\n",
    "    introns.append(full_sequence[last_pos:])  \n",
    "\n",
    "    import pandas as pd\n",
    "    exon_intron_data = pd.DataFrame({\n",
    "        \"Type\": [\"Exon\"] * len(exons) + [\"Intron\"] * len(introns),\n",
    "        \"Sequence\": exons + introns\n",
    "    })\n",
    "\n",
    "\n",
    "    print(f\"Exon and Intron Sequences {exon_intron_data}\")\n",
    "\n",
    "else:\n",
    "    print(f\"No sequence data found for {target_transcript}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0e0aff3-ef70-4e43-a25e-ce222508debc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ Comparing exons for transcript: chr1:901862-911245\n",
      "\n",
      "Exon 1 [901994-902084]\n",
      "Original : GGTGAGCGGGGCGTGGGTGCGGCCACCTGGGCGCAGGGCTCCCCCACCCGCTCCGGGGCCAAGCCACGAGACCCCTTGCCTTGTCCCCAG\n",
      "Shuffled : TCCGCCAAGGGTTCCGAGACCCCCCCCCGCGCGGGCCGGCAGGTGGGGCCCTGGGGCCCAGACGCCACCGCGTTGCCTCGAGGTCAGTGT\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 2 [902183-905657]\n",
      "Original : GGTGAGCGCGGCGTGCACGGTGGCTGTGGTCTGGGAGCGTGGCTCTGCCCGCGCGTGTGTGCCGTGTGTCCGTGCAGCTCAGGGTCTTCCCCTCGCCCCCGGGGCGTTCAGACCCCGGTAGGTGAGGAGCCGACGCTGACTGCCCCGCCCTGGGGAGCTCATCCGGCCGAAAGTCTGAGCGGAGAGCGAAACCGCGACGCAGGACTGAGTCAGTCGGGAGAGAGCTgggcggggcgggccgcgggcggggccTCCCGGGACCTCCTCCCGCAGGGGACCAGCGCACGTGTCTCTGTCGCTGCAGTGGGGCTTGGGGTCGTGCCGGCAGGTAGTGTCACTTGAGAGTAAGGGGCCTCAGATGAGCAGCGAGCCAGCCGGTCACCTACACCGGCCTTTCTCCCTGGGTGTCCCGCAGGAAGTTGCCGCTGGGTATGGCCGCCTGGTCCACAGCAATCCCTCTGTGCACGCGCCGGCCTCCTCCCAGCTTGCCAAGGCCATGCAGCAAGGAGGTGTCTCCCAAGCACACACTAGGCCACCTCTCCGTCAAAGGAGCAGAAGACCACCCCCATGCCCACCCCCCCCCCAGCCCCTGTGGGACTAAGGGTCTTCACAGGGTAGATCCCAGCCCCTTTCAGATGTGTCTGGTGCTGGGATGAGGGAACAGGACCAGGAACCTGGGCTTCAGGGCAGACAGGAACCCCCACTGCACCCCCGTCAGGCAGCCTGGATGGGTGGGCTCCGCTATGTGCAGGCCCTAGGAGCATCTGCGGGAGAGGACGCCTGCCCTGCACTCCTGGGTCCAGCAGCTGGCATGGCCGTGCAGACCCCTTTGCCAACATGCCCCCGCCAGGCAGGAAGCTCGTGCCATACCCTCTCCTGACCTCCAGAGTTTCAGGGCTCCCTGCAGACAAGTACTGGTTTCGGAAACCACGCTGCCTCAGAGGAGTGGGCAGGGAGAGGGAGCCGGGCTGGGTTAGGGGTGGGGTCCCTGTGGCTGCATCTCAGACACCACTCCCTCTGTCCAGGCCCAGGACATGCAGTGTGACCCTGGGGGGTCCTATGAGGTGTGGTGGGTGGCCGTAGTGGTCCTCCCTTGCCAGGACATGGTAGCCACAGAGGCAACAACCTGCTGAACCTCGGGAAATGGGGCTGGGGGAGGAGCTCTGGGCCAGTCCTAGAGCCAGTGCTCCGCGGCCCTGTACCTGGGCTCTGCACCCAGTGCTGTGAGCTGGCTCTACTACGGCTCCAAGCTGAGTCCTCCCCGGGGCCCCGAGGCTGAGCTGCGCTGCAGGGCCAGCTGTGTGGCCCTTCCTGGTCTGTGGCCTATTTTTCATGGGTGCCAACCCGGCATCAGTTCCCACGCTGGGTGTGGGTACAGGAGGCCCCACCCAGAGGCAGAAGGCCGGGAAAGCCATGCGTCTGTCTGTTCACATTCCTGGGCCTGCTCAGGGGCAGCTAAGCTTGGCCAGGGCACAGCAGCTCTGCAGGTAGCTCTCGCCAGCTCCTAGAAGGAGGAGGACACCTTCCCACTTGGCCCTGAGCCTGTGGATCACAGAGGTCTCCAGCCCCAGGTTCACACCCGTCCCTTTTCTGACTTCCAACCTGGTGATCAAAGCCTGCATTTGTAGAGGACTGTGTGTGCCAGGATTGACTACATGTGGCGTGGATGTGGAGTCCCCTCCCCTCTCAGCACATGCAGTGATACCCACAGCACATCccaggccctgtcttaggctctggggacacagcagtgaacggaatagacaacccctgttctgctagtgctcccattcccacggcagagacacacgacgggacaaatcaatagatccagtgagcggggtcaggtcggtgctaaggagaaaaacagaggggatgagtgcagaggatcagggaaggctggaattttagaaaggctttcagaaaaggaattcttggagcaggtaatgtctgagcaaagctgcaaagcgggtaaggaggggggtcaggactgagcggtccccaaagcaggaacagaacatgcagcagccagagagcatgggtctgatgtgaggagacagtgaggccggggcagcCACAGAGACCCCTCGGGGGAGACCCCTCGGGGCACTGCCAGGACCCCGTTGTCCCCTGAGTGAGGCATACGGAGCGGCCAACACACAGAGGATCGACCCCAGAAATCTCCTCACATATTtcaggctgggggctggaaatgggcaacggtgaggacactgacaccgaggagattggcaaacgttacaaatcagggcatcggtgtgtttctgtgtctgttttttgtagaaccaggtgttaagcatttaccatacaatactgCAGACGTGGGAGGGTTTTGAGTGTTACCCAGACACCCCCAAATTCTTCTAACAAGTCTTCAGAGGTTGAATAACCCAAGCCATGTTACACAGCTTAGGAAAGAATGAAGATGGGTTTAATCCACTGAGTCCAGATCTCACCATCTCTACCACTAGACCAGGCAGCTCTGTGTGGGGAAgtgtgtctgtgtgtgcgtgtgtgtgcacaagtgtgtatgtgtgcatgtgtttgcacgtgtatatgtgtgCACACATCTGTGTGTATATGCACAGgtgtatgtatacgtgtgtgttggtgtatttgtgtgtgtgtggaaatgtatgcatgtatgcagtgtgcatgtgtgcatgcatgtgtgtgcatgtgtgtataggtgtgcatgtgtatgtgtgcacgtgtgtatgtgcacgtgtgtatgtgtatacatgtgtatgggtgtatttgcgtgtgGAtatgtatgcatgtatgcacgtgtgtgcatatctgtgtatgtGCAACtgtgtatgtgtgcacgtgtgtatccatgcatatgtgtgcgtgtgtgtacatgtgtgtataagtgcgtatgtgtgttcccgtgtgtatgcatctgtgtctgtattgtgtggcatgtatgtgtctattgtgtgtgtatgcatgtgtgcgtgtatgcatttatacatgtgtatgtgtgCgcatatatggatgcatgcatgttgcctgcatgtgcacgtatcgggtgtgtgtgcgtgtacgtgtacacgaatatgcatgtatgcatatatgtatgcacgtgtgTGTAAATTCTGCCGTGACTGTTTTCCTGTGGAAAGGCTGAGTGTGCAGGTTCTGTGCCTGTGGGGGGCTGTTCTTCACGTATGTGTTGTGTGGCTGTGCACAGGTTCTGTGCCTGTGGGGGGCTGTTCTTCACGTATGTGTTGTGTGGCTGTGCACAGGTTCTGTGCCTGTGGGGGGCTGTTCTTCACGTATGTGTTGTGTGGCTGTGCACAGGTTCTGTGCCTGTGGGGGGCTGTTCTTCACATATGTGTTGTGTGGCTATGCACAGGCAGACCATGCTATAGTCCTGTAGCTGTGTGGATGCGAGCGGAGGGGGTGGGGGGCCCAGGGGAGGCCCCCTCCCCTGAGCTCTACTCCTCCTAG\n",
      "Shuffled : agaACCCGAGTCgtggtaTCCtcgATCCTAGGGtgtGAGGGAgatggtCTCGGTTACCCCCCCAGCtccCACattCCTGTGgcCatgaaaAGCCCTgCACCGGCAGGAGGCcgtATCATCaacGCAGATTAGCTGCCTatcACAatatatGGGACATGTtatgatACAGCAtagCTGAGTGGGAGGTTTTGTgggTGCAGTTTCgcaTGACGTGGGagcTCTGTGACGgtgAGGtgtCGGtatCAAcgaTGCTGGCCGTCAaggGTGGCTTGTGCCgcaATGCCCGGCGGTATCACActcGAGGGCGTGcggactgggACCTACTAGTTCCTGttagcaCCTGTGgaacacGCTTCAgCgCTAGCAGCAgtaCCGATGTGGGTACCTTCCGTAaatTggAGTGTGCCTtagTGTcagtgtAGTTCCgtgtgtAGCgtggcgtGCCAGTTCGTGAGTACAgtgtccTATACACCCCCTggcGCCCTGTCCcatGTCAGGGGTGCCTCCAGAGCAttagtaatatgtGTGGCTTGGACATGGCCCcactgtCCCGGGatgGGGAGGCGCTGGAGACGCAGGGTGAGTCTGTGTgaaGTGCTCgtggtatcgtgcgagCCCCCTgtaGCCttaCCCGAGggaACTCTGgtaGCTGGAaattgtggcCCAGGGtacAACGGGGATTGGatgACAtgtcatgaaGGCcacgtgTGTCCGagcataCTCTGCGTGGCCAAGTGGGGCCCTtgtatgCGGGATgtgCccCAGCATaggAGGTTCCCTTCACGGGTTGAGAGGAGAGGCTGTTCCgtcTCGTTGGAGGGGCCGCGTCTGtcgTGCcaccagTCCggagtgGCTTTCAATCGTGCTgtatccCAGAGGTGTAGCTCTGAGgctTATCCTGCCGGGcaaaacGAGgtgCCACTCAACAGCGCTCCCCGCGGTCTCGAACAGTAGaatTGGagcggaGCAccaCAACTCTCCTGCCGGCTGTTTCTGAGCCGCatgAGTAAGTCTatgTTACAGCATaggAACtgcGGCGCTgccCTGCCTctgACCagaCCTtatTGGAgtACAAGGTGGGAAcggCGAGGGTGCTGGAGGCGCgtgtgcAGTGGTTGACTCCCCgggCCAcatCCGCCAtgatgcTTAACAtgcatggagGGAgtgcaaAGCGACtacCACGACTtcCCCTCAtattgcgcgATTtgacgtGGGtccGAGTGGCTGTACCCACCGAAACCTtctGGAGCTcatCTGACAGGCACCGTGtcaCCACCGgggAGGCCAGTATGCAACcccacaaggtctACTGGTAAGGCAggactgtgtaaaGCCtaaaagTCTtatGCCacaGCTGACGCCGCATGTTGGtgcCAAACCCAGCCGagttgtgcagtgcagGAGtatTTTataGGGTCAGGGtgtGGACTTCCCCTCTGTtgtTGTGGCTGCgtcGTCcggtgcTGTacaTGGcggGCTGCCCCTGGAACCCTTACGGAGTATGGGCAGcagGCAGGAatgcggcagTTCGGAAACttgTTCCAGcacAGAGGCGTGGTTATCtgttatCCACCTCTGaggCTGCTGgtgCCTGAGAGAgcgGTGGCAtgttgtGCTCAGGGCCTGTCAGCAgatGATATAagaCGGagggacggaCCAaaaACTCTAaggCCATCCGTCAGCatcGAGTGTGGGTGGTCTAGAtctCATACGTGTCGAacgTAAgcgGTGatgTTTGGGGTGtgtCGTAGGCAGacgccaAAACCGGACGTGGTGATGtgtTCCCCTgggcatgtgCAAgtggtaggttggACCgggTGGATGGTTGGTCCCcactttTTTTCACCGGGCAGGGCCtgggtggtgTGTGGCgcaCTCGCCGCCtgtCTCgggCCTTGCAGGTGTTAGGAGGCCCCGttgACTGGAGGGTCCGCTTGTgtgcaggggACACCGGAATCTAGCGCATGCCCAGAAGATgtcCAGCACCCTTTCAGActgGCCCTGgtgGGAATGCCGCCACGTCCTGACagtCTCTTCCGCgctTCAcatTTGggtCTCCTCGGTTGTtgtCAGCACAAAAGATAAACCgtgGTCCTCACTCACagcGGCAGCCGCCTGTGTCCGACCACAtatGCACGCagtcacgcgCCATACctgCAGcttGGCGTTgcaTGTACGtatGCAGCACCAACTGAtGGCtgtagcTGTgtgtctCCCgtgGATaggGTCGTTGTCACTGAGGCACCAataTGGGCTAACGTAgtgGCCTCGaaaATTTCTCTGttcatgttgGTGCCGACACCACTCGGTTGCatgtttatgTACTGGATGCTCCTGAGGTGCATAGTCagtGGAGAGtgcGGCCTGtagAGTGGAACCGTCAGGGAGACACTGtggCAGGCCGCAcgaGGACCTGTTGTTGACGTGgtgTTTGTCcaactggtgGAGACTTGGtcttgtAAGtgtCACcagTAGccaTCAGGTcggtgtGCACCTtatGTGCCTGTGGTTACCtgaAGGGTTttgggagcaCTAgcatatGGGTGTCCCGACCCCgagTGGTTTGGTGCCaggatgGTTCCTGAGAGAACAACCAGCTGGGTGatgtatAAGtgtCCCacaGCGGGCAGGGGGCCCTGGGGGGAGATCAGAACCGCCTGGgcgCAGTGCgtgACCtatgcagtgatatgttgtGCTCCAtgtCCCttgGTGgagCCAGGGCTGTTTcacGGTGTTTGTGCACTGatgcccCTTgtgAAGCACtatacaGGTCCCGGAgtgTGTGTGGGGCCCgtgTTGCGGtatAGACATaaaCACTATTTCGCCtggaggCACTGATCAttgTGTctgtgtCTCGTGGGGACAGGTcgaCTGagcCCCtatgcaCTGCTGtgtTCCcattgctgtgcaGGACTCcatggcctgCCAGCActgggcCTGgtgatgGCCCAGGAGAGGAATgtgGCTCCAatgCCTtgtCCGTCCTTCGTGCAGtgcatgCAGTGCCTGGGGgtgtttGAGGTCAGCgcacgtGTCGGGtatCACacgGCGaatGCCGCCAGCaggGTGGCAGGAAGCGGGCTCTGGGGAaggCTGacaAAACCCGGTtgtTGCTTGATGgaccatattGTGAGGCAATGCTGTTGCAGCGCCCTCAGgACCGGGagtTCCCATGTGCCCAGTGTGCACcaagcaGGGTTCtgtACCAGGgagacgGAAgccgtaAGGCCCCTCTGTGGAGCCACAGGCCGTgcaaagCTAtggAGCGCTACGACCCCTGCTTCTATGgccAGTAGACTGgacAAACCCCCACAGTCTGGGatgggaGGACTCgtgAGGCTCTGCCTTGAGtgtatgCCCcgtgtgTGAatggtgACACCGCCAAATGGGTCCttcATGctaAGCtgtCCGGCGATGTAGTGGcatGCTtgcTGTGGGGGGCGGtgtATAAGGCAACCAATCTCTCCCtgtGGC\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 3 [905803-905901]\n",
      "Original : GGTGGGGGCCGGGCTGGGTGGAGCACGCTAAGGGTGCAGCATCCCCATCAGCCTGGGGCTCCCCAGACTCCGCACTGACGACCCTGCTCCCCGCGCAG\n",
      "Shuffled : TGACCCACGCCCGGGTCCGGGAGCCATGGCCAGCGATGCCCTCTAGCTCCGCAGGCTCGGTGGCCCCCGGTGACTCACCGCAGGCATGGGGGTCAGAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 4 [905981-906066]\n",
      "Original : GGTGAGGCGGTGGGCAATGGGGTGGGGCCATGGCCGCCCTTCCCTCCATGGATCCCTGAAGCTCCTCCTACCCTGTGCCTGGCAG\n",
      "Shuffled : GGTGAGGTGTGCCTTGGATGCTTTCCGGCCCCGGGAGCGCCGATGGTGGGGCCTCCGCCGGCATGCCTTGATACCAACCCTCCAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 5 [906138-906259]\n",
      "Original : GGTGTTTGGGATGCTTCCCGGGCCCCCAGAGGCACTCCTGACCCAGGACTTGGAGAGGGGCCTGCCCTGTGGCTGCGGAGCACGTGTGTGTATGTGTGTGCCCTCTCTGCCCTGCCCGCAG\n",
      "Shuffled : TGTTGTGACCCTGGTGGGACGGGACCACGGTGGGCCCCCCTTTCTCTGGCCGCTACTTATCGGCCCGCCGTGGCCAGACTGGATGGAGTGGGCGAGAGCCCTTCCTGGGTTGCACTGCTCG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 6 [906386-906493]\n",
      "Original : GGTCAGTGCCGGGGACCCCACCCCCCTCCCCACCCTGATCCTCGCAGCCGGCTCTGACCTCCTCCTCACAGGGGTCCTGCGGAGACGAACTCCCCTGGACTTTGCAG\n",
      "Shuffled : GGTCCCCAGAGGCCTGCCCGACCCCGCCCTCCCCACCGACCGGGGGGCCCTGGCACTCGGGAGTCTGCTCTCTGCAACTACACTCCTCCGAGTCGACATCTTTACCG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 7 [906588-906704]\n",
      "Original : GGTGGGTGGGAGGTGCGTGGGGCTGTAGGGGGATGGGAGGGGTGCATGGTGGTGGGCAGGGCGGTGCAACAGCAGCTCAGTTCCCTCATCGCCTGACCCCACCCCCACCCACCCAG\n",
      "Shuffled : CACGGGCTCAGTCCCGCCGTGGTGTGCCTCGGGGGTTCCCGTCACATGGCGGAGCAGCCCTGGGCTGGAGTAGAGCCCGGCCAAGGGGTGATCCCAGGGTGGTGACAGAGGGTGAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 8 [906784-907455]\n",
      "Original : GGTGGGCCCCTCCCCACTGTGGGCCCGCCCCAGGGAGGCAGCTGTGTGGGGGTGGGAGGGGGCAGCCATGCAGGAGGAACCTGTATTTCGGGGCTTGTTGAGACAGAGCGCAGGGCCCTGCCCACCCAGCCCGCGGTCCTGGTCAGTCTGGTCTCCGCTGTGACATGGAGGAGCCCCTGCCCGCCCTGAGGTTCTCACCCTGAGGTTCTTGACCGTCACCCTGGGGCGGGCAGCATGGGGCTGAGAGCCCCTCCCGGGGACTCCCTTGTGTGTGCCTGCCCGTGAGGCGTGTGTGTGCACACGTGCGTGTGCGTGTGTCCTCCAGGGCGGGCTCCGGGAAGGGGGTGTGCACTCACGGTAACCTTCAGTCACTGAGGAACAAACACAGGGCCCTCCCCATGGTTCACCCGGCCCAACTTCTTCTCTGGGGACCCCAAGAGGCCAGTCCTGTGGCTTTGTGTGTCTGCCCCTAGCCGGCGAAGGCCCCCGCGCCAGGGTTGGTGGTAGCCCCAGGAGACATGGGACCCACTGGCCTCCCTCAGGGTGGTGGCCTCGGTGCCCAGCTCTCCAAGTACCCAGCCAAGCTCTGGAGGGCAAGAGGGTGGGATGAGGCCAGGCGGGGCCCCGGGGCTGCCCCTACAAAGGCCTGGCCCTCAATCTCTGCTCCGCAG\n",
      "Shuffled : GGTGGCTCTCTTCACCCAACGTGTCCGCCGGGCCTCAGGCCAAGCTCGGGGGACAGCAGGACCGGGTGCAGCCTAGACAGTGGGGAGCGAAGATGTGGGGGGTGGCCCCTGCCAACAGGGCCGGGCAGCGCTGACCTCACCACGGCCCCGCTACTGGAGATGCCCCCGGTCCCGCTAGGTCAGGTGTAGTGCTCAGGGGGGTGGCCGGCCTGACTCCACGTTCCTGACCACACGGCCCAGAGCCTGGTCCGGCCTCCGTGCCAGCGGCGGAGTTCTCTGGGCCCGTTATGCACAGAACAAGGCAGCAGTCTCCTGGGGGGCCCTGTTCCATCGCAGGATGTGGGAAGCGTCAACCCCTGGTGGGTCCGTGTGTGCAGGACCTGCCGTGCCACAAGTTGCTTAGGTGTAGCAGGCCTCACGGCTATAGGGCCTCACGGAGGAGGGGGGCTTTGTGCTGTCTTCCAGCTCTGGAGGGTCTCGAGCAACTCTACAGTTGTGTGCTCGGTGAGGGCGGGGCTGCACCATCAATGCATTGTGTCCGCCCCGGGAGGAAGTGCGCCTGGGACCCGTCCGGTCCCGCCCCCCTTGGTTGCCCCTGACCCGCCTAGGGGCCCGGCCTCGCGCTCTGGCCCGCCCTGAGGGGTTTCAACGCCGCCTGGAAACTGCGTCCC\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 9 [907530-907668]\n",
      "Original : GGTAGGGCCCTGACCCTGGTTCTGCCTCCCGCCTGGCCAGGCCATGGTGGGGCGGGAGCCTGGGGGACGCCCGACTCTTTAGTGGGGGCGCTGAGGGGTGCACCCCCCCGCCAGCCCCTCACAGCATCTGTATGCCAG\n",
      "Shuffled : GTAGCTTCCCTTCACTCAGTCATGCGCCTGACCCGTGGCCCGCGTGGGGGTAATGTGGCCCCGCGGGCGGGGGTGGCCGCCTCTCGAGGGCCATCAGCCCCTGCCTTGATATCACCCCCCGGGGGGGGCCAGCGGAAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 10 [907804-908241]\n",
      "Original : GGTGAGGGGTCAATAGGCCCCACAGCCCAGGTCCTGGGCAGTGGTAAAAAGGGGGCAGCAgaccgggcgtggtggcgcacgcctgtaatcccagcattttaggaggctgaggcgggcggatcatttgaggtcaggggttcgagatcagcctgaccgacgtggagaaaccctgtctctattaaaaatacaaaaaaattagccgggcatggcggtacatccctgtaaccccagctactcaggaggctgaggcaggagaatcgcttgaccaggaggcggaggttgcagtgagccgagatcgtgccactgtactgcagcctgggcgacagaataagacttcgtcccaaaaagataaaaGGGGGGACAGCACGGTGGGTCCAGGCAGGGGCGGGTGGGGAGGGGGTCCAGCCCTCACTCACCCCCACTCACTGCCCATCCAG\n",
      "Shuffled : GacatagcccaggCAGttccgagcaggcGCAtaaCCCtcaGCCACTGGGtaatgcctatcacagGGGgcgagactgAGGataggtCAGGGTGGAagcCACCGGccagggCAgAGCaattggctgCATtacccgccttccccaaccgaggcgggagttcacggcGTGgaaCAAacccatctgaactttatcgacaaaactGGGgtaCCAGGGcccCACaaGttgtgtatgCCAgtcaaagcgtgcgcaaggAGGGGTccgagtTCACCTTCCaggggcttatcgAAAGGGggcGGGcgtagatctggacgtAGGcatTCCGTATCCgaacggCAGTAGCAGtgaaggAGGggcAGGgccgactgagggGCCcgcgataggtgtCCCagacaaggagctTGGCCAgatcgataacatCTCGTGaaaggaggtGTGcagG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 11 [908390-908566]\n",
      "Original : GGTGAGCGCTTACGGGGTGGCAGACGAAAGTGGGGCAGAAGGCTGTCGGGTAGGTGTGTGTTGGGGACCCTGGTTCCTCAGGGGAACTCAGACTGGAGGGAGCCCCTTGCAGCCTCAGAGAGTTGCACGGGAGAGGGGCCAAAGGGCTCTTCCTGGAAGGTTTGTGGTCCCCACAG\n",
      "Shuffled : GGTGTGCCCAGGAAGGGTCCCAGATCGCTGCCTTCAGGGCGGGCCTGGAGACACAAGACCACGCGAAGGGACGAGTGGCTTTGGTCCGTTGAGCTGAGGGGTTCAGGCCAGTTAGGATGCGTTAGTGAGGAGTGTGCAGGTGGCGGGAACTGTCCAGGGGGCTCTCGCTTGAAGAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 12 [908706-908880]\n",
      "Original : GGTGGGCCCAGCACACCCACACAGCCCCTGGCCTGGTTCCCACCGTTCCGCACCACTGGTCTCTGTCTCTGGGCCCAGTTGTCCTGGAGCCACCCAGAGGCCTCTTGCACAGAGAAGGGGTGGCCTAGGCTGTTTCTAGCCGAGAAGCCCATTTCTCCCACCTCTGCCCTGCAG\n",
      "Shuffled : GGTGGGTGACAGTTCATTGGGTTCTGTAGGGCCCCCTGGTTCTCAGACGAAGACACTCCGCCTTGGCTCTCGGCTCTCCTGCACACCTGGCCCACCTGGGGGGCACCCCGACCGAGAGCTTCGTCCACTCTCCCCAGGAACACGCCGCCCCAGTTTGCCCTTCTCCACACGAAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 13 [909020-909213]\n",
      "Original : GGTCAGGCCCCTGCTGGGTGACAGAAAGGGTGGGAGGTGCCTGCAGCCTGAGGCTGGGGAGGTCTAGACCGTGCGTCTCACCCTGGGGTCTGGGGCTGCCCCAAGCCTGAGGTCTGTTCAGGCTCTGATGGGAGGTTGAGGTTCTGGGGGCCCCTGGCTGCCACCCAGGCCCCAGCCCTGGCTCTCCCTGCAG\n",
      "Shuffled : GGGGGCCACAAACAGGCAGGTGAGTGGGATCAGGTGCCTGCTTCGTTGAGTCTGTCTGCAGGCTGCACCTGGTTAGGGGCCCCGTGCCTGCCCTGCTGAGCGGTGGCAGGTGGCCCGGCGAACCCCCTAGCTGACTGGCCTGGTGCCTGCCCTCTGAGGGGTCATGGCCCTGCCTCGGTCCGAGATGCGGGAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 14 [909431-909696]\n",
      "Original : TGTGAGTAGCAGCCCCCACGCCCGTGTGCCCCGGGCTCCGGGCTGGCCGGGGGTCTGGTGTGGGGCCTCTTGGGACTCTGAGGGAGCAGGGAGGGAAACAGGAGGACGGGGGCAGATGGAGGCCAGGGGGGCCAGTAGGGAGTTGGGGAGATGGGACTGGGGAGGGAGAGCAGGGGGACATGGGGGGCTGCACAGTGACAGGCCGCCTCCAAGCTCCCTGCCTGGGGCTGTGCCCGGCTCTCAGACTTGCGGTTTGGGGTTCCAG\n",
      "Shuffled : CCGTCATCCACAGGGGAGGGGCTCGCCGGGAGGCTCCAGAGCTTGGACGCCCTCTGGATGTGTGGGGGCGCACAGGGGCCAGGGTGAAGGGGGTAGTAGGGACGGGCCAGGGGGAACCCTCAGGCCTGTGATGAGTGGGAGGGACAGTGTGGACCGGGCAGGGGGGCCTGGGGCGCTTCCGAGAAGGGCAGCCCCGGCCAGACGGTGGCAGGGAGGCGGGCCTGACTGCGCGGACTGCTCGTTCTGGGTTTTTTGGTTGGACCAG\n",
      "✅ DIFFERENT — exon was shuffled\n",
      "\n",
      "Exon 15 [909744-909822]\n",
      "Original : GGTGAGTAAGGATCCTGCCTCCTGAGGTGAGTGCCTGTTGCCTCCCACAGGCTGACACATCTCTGCCTTCCCTACCAG\n",
      "Shuffled : GGAGCTGAAGTGATCCCCTACAACCGTGTGCGGTTCTCACGGCTGATTGATCGATAGTCCTCCCCCTTTCGCCGCCAG\n",
      "✅ DIFFERENT — exon was shuffled\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "\n",
    "target_transcript = \"chr1:901862-911245\"\n",
    "canonical_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_sequence_unflanked_unshuffled.txt\"\n",
    "shuffled_file = \"/mnt/lareaulab/sdahiyat/illumina/permuted_exons_by_codon.txt\"\n",
    "junction_file = \"/mnt/lareaulab/sdahiyat/illumina/canonical_dataset_created.txt\"\n",
    "\n",
    "def load_sequences(file_path):\n",
    "    sequences = {}\n",
    "    with open(file_path, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if not line or \"\\t\" not in line:\n",
    "                continue\n",
    "            header, sequence = line.split(\"\\t\")\n",
    "            sequences[header] = sequence\n",
    "    return sequences\n",
    "\n",
    "original_sequences = load_sequences(canonical_file)\n",
    "shuffled_sequences = load_sequences(shuffled_file)\n",
    "\n",
    "def load_junctions(file_path):\n",
    "    junctions = {}\n",
    "    with open(file_path, \"r\") as f:\n",
    "        for line in f:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if len(fields) < 8:\n",
    "                continue\n",
    "            chrom = fields[2]\n",
    "            start = int(fields[4])\n",
    "            end = int(fields[5])\n",
    "            exon_starts = list(map(int, fields[6].split(\",\"))) if fields[6] else []\n",
    "            exon_ends = list(map(int, fields[7].split(\",\"))) if fields[7] else []\n",
    "            transcript_key = f\"{chrom}:{start}-{end}\"\n",
    "            junctions[transcript_key] = (exon_starts, exon_ends)\n",
    "    return junctions\n",
    "\n",
    "junctions = load_junctions(junction_file)\n",
    "\n",
    "def compare_original_and_shuffled_exons(transcript_key, canonical, shuffled, junctions):\n",
    "    if transcript_key not in canonical or transcript_key not in shuffled:\n",
    "        print(\"Transcript not found in one of the sequence files.\")\n",
    "        return\n",
    "\n",
    "    if transcript_key not in junctions:\n",
    "        print(\"Transcript not found in junction file.\")\n",
    "        return\n",
    "\n",
    "    original_seq = canonical[transcript_key]\n",
    "    shuffled_seq = shuffled[transcript_key]\n",
    "    exon_starts, exon_ends = junctions[transcript_key]\n",
    "    seq_start = int(transcript_key.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "    print(f\"\\n Comparing exons for transcript: {transcript_key}\")\n",
    "    for i, (abs_start, abs_end) in enumerate(zip(exon_starts, exon_ends)):\n",
    "        start = abs_start - seq_start\n",
    "        end = abs_end - seq_start\n",
    "        if start < 0 or end > len(original_seq):\n",
    "            print(f\" Skipping out-of-bounds exon {i+1}: {abs_start}-{abs_end}\")\n",
    "            continue\n",
    "\n",
    "        original_exon = original_seq[start:end]\n",
    "        shuffled_exon = shuffled_seq[start:end]\n",
    "\n",
    "        print(f\"\\nExon {i+1} [{abs_start}-{abs_end}]\")\n",
    "        print(f\"Original : {original_exon}\")\n",
    "        print(f\"Shuffled : {shuffled_exon}\")\n",
    "        if original_exon == shuffled_exon:\n",
    "            print(\"SAME — not shuffled or identical by chance\")\n",
    "        else:\n",
    "            print(\" DIFFERENT — exon was shuffled\")\n",
    "\n",
    "#  Run the pipeline \n",
    "compare_original_and_shuffled_exons(\n",
    "    transcript_key=target_transcript,\n",
    "    canonical=original_sequences,\n",
    "    shuffled=shuffled_sequences,\n",
    "    junctions=junctions\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5afcd935-48ae-484f-9884-1cdb74ffa1f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (september)",
   "language": "python",
   "name": "september"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
